{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import biomartpy as bio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biomart</th>\n",
       "      <th>version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSEMBL_MART_ENSEMBL</td>\n",
       "      <td>Ensembl Genes 88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSEMBL_MART_MOUSE</td>\n",
       "      <td>Mouse strains 88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSEMBL_MART_SNP</td>\n",
       "      <td>Ensembl Variation 88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSEMBL_MART_FUNCGEN</td>\n",
       "      <td>Ensembl Regulation 88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSEMBL_MART_VEGA</td>\n",
       "      <td>Vega 68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                biomart                version\n",
       "0  ENSEMBL_MART_ENSEMBL       Ensembl Genes 88\n",
       "1    ENSEMBL_MART_MOUSE       Mouse strains 88\n",
       "2      ENSEMBL_MART_SNP   Ensembl Variation 88\n",
       "3  ENSEMBL_MART_FUNCGEN  Ensembl Regulation 88\n",
       "4     ENSEMBL_MART_VEGA                Vega 68"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bio.list_marts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "datasets = bio.list_datasets('ENSEMBL_MART_ENSEMBL')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>description</th>\n",
       "      <th>version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>hsapiens_gene_ensembl</td>\n",
       "      <td>Human genes (GRCh38.p10)</td>\n",
       "      <td>GRCh38.p10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  dataset               description     version\n",
       "31  hsapiens_gene_ensembl  Human genes (GRCh38.p10)  GRCh38.p10"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human = [ d for d in datasets['description'] if 'Human' in d ]\n",
    "human = datasets[ datasets['description'] == human[0] ]\n",
    "human_dataset = human['dataset'].to_string(index=False)\n",
    "human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ensembl_gene_id</td>\n",
       "      <td>Gene stable ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ensembl_transcript_id</td>\n",
       "      <td>Transcript stable ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ensembl_peptide_id</td>\n",
       "      <td>Protein stable ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ensembl_exon_id</td>\n",
       "      <td>Exon stable ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>description</td>\n",
       "      <td>Gene description</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chromosome_name</td>\n",
       "      <td>Chromosome/scaffold name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>start_position</td>\n",
       "      <td>Gene Start (bp)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>end_position</td>\n",
       "      <td>Gene End (bp)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>strand</td>\n",
       "      <td>Strand</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>band</td>\n",
       "      <td>Karyotype band</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>transcript_start</td>\n",
       "      <td>Transcript start (bp)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>transcript_end</td>\n",
       "      <td>Transcript end (bp)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>transcription_start_site</td>\n",
       "      <td>Transcription start site (TSS)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>transcript_length</td>\n",
       "      <td>Transcript length (including UTRs and CDS)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>transcript_tsl</td>\n",
       "      <td>Transcript support level (TSL)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>transcript_gencode_basic</td>\n",
       "      <td>GENCODE basic annotation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>transcript_appris</td>\n",
       "      <td>APPRIS annotation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>external_gene_name</td>\n",
       "      <td>Gene name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>external_gene_source</td>\n",
       "      <td>Source of gene name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>external_transcript_name</td>\n",
       "      <td>Transcript name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>external_transcript_source_name</td>\n",
       "      <td>Source of transcript name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>transcript_count</td>\n",
       "      <td>Transcript count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>percentage_gene_gc_content</td>\n",
       "      <td>% GC content</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>gene_biotype</td>\n",
       "      <td>Gene type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>transcript_biotype</td>\n",
       "      <td>Transcript type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>source</td>\n",
       "      <td>Source (gene)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>transcript_source</td>\n",
       "      <td>Source (transcript)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>status</td>\n",
       "      <td>Status (gene)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>transcript_status</td>\n",
       "      <td>Status (transcript)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>version</td>\n",
       "      <td>Version (gene)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1447</th>\n",
       "      <td>uniprotsptrembl</td>\n",
       "      <td>UniProtKB/TrEMBL ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448</th>\n",
       "      <td>cdna_coding_start</td>\n",
       "      <td>CDS start (within cDNA)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1449</th>\n",
       "      <td>cdna_coding_end</td>\n",
       "      <td>CDS end (within cDNA)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1450</th>\n",
       "      <td>5_utr_start</td>\n",
       "      <td>5' UTR start</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451</th>\n",
       "      <td>5_utr_end</td>\n",
       "      <td>5' UTR end</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1452</th>\n",
       "      <td>3_utr_start</td>\n",
       "      <td>3' UTR start</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1453</th>\n",
       "      <td>3_utr_end</td>\n",
       "      <td>3' UTR end</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1454</th>\n",
       "      <td>ensembl_transcript_id</td>\n",
       "      <td>Transcript stable ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1455</th>\n",
       "      <td>ensembl_peptide_id</td>\n",
       "      <td>Protein stable ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1456</th>\n",
       "      <td>transcript_biotype</td>\n",
       "      <td>Transcript type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1457</th>\n",
       "      <td>strand</td>\n",
       "      <td>Strand</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1458</th>\n",
       "      <td>transcript_start</td>\n",
       "      <td>Transcript start (bp)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>transcript_end</td>\n",
       "      <td>Transcript end (bp)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1460</th>\n",
       "      <td>transcription_start_site</td>\n",
       "      <td>Transcription start site (TSS)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1461</th>\n",
       "      <td>transcript_length</td>\n",
       "      <td>Transcript length (including UTRs and CDS)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1462</th>\n",
       "      <td>cds_length</td>\n",
       "      <td>CDS Length</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1463</th>\n",
       "      <td>cds_start</td>\n",
       "      <td>CDS start</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1464</th>\n",
       "      <td>cds_end</td>\n",
       "      <td>CDS end</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1465</th>\n",
       "      <td>ensembl_exon_id</td>\n",
       "      <td>Exon stable ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1466</th>\n",
       "      <td>exon_chrom_start</td>\n",
       "      <td>Exon region start (bp)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1467</th>\n",
       "      <td>exon_chrom_end</td>\n",
       "      <td>Exon region end (bp)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468</th>\n",
       "      <td>strand</td>\n",
       "      <td>Strand</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1469</th>\n",
       "      <td>rank</td>\n",
       "      <td>Exon Rank in Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1470</th>\n",
       "      <td>phase</td>\n",
       "      <td>start phase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471</th>\n",
       "      <td>end_phase</td>\n",
       "      <td>end phase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1472</th>\n",
       "      <td>cdna_coding_start</td>\n",
       "      <td>cDNA coding start</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1473</th>\n",
       "      <td>cdna_coding_end</td>\n",
       "      <td>cDNA coding end</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1474</th>\n",
       "      <td>genomic_coding_start</td>\n",
       "      <td>Genomic coding start</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1475</th>\n",
       "      <td>genomic_coding_end</td>\n",
       "      <td>Genomic coding end</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1476</th>\n",
       "      <td>is_constitutive</td>\n",
       "      <td>Constitutive Exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1477 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 name  \\\n",
       "0                     ensembl_gene_id   \n",
       "1               ensembl_transcript_id   \n",
       "2                  ensembl_peptide_id   \n",
       "3                     ensembl_exon_id   \n",
       "4                         description   \n",
       "5                     chromosome_name   \n",
       "6                      start_position   \n",
       "7                        end_position   \n",
       "8                              strand   \n",
       "9                                band   \n",
       "10                   transcript_start   \n",
       "11                     transcript_end   \n",
       "12           transcription_start_site   \n",
       "13                  transcript_length   \n",
       "14                     transcript_tsl   \n",
       "15           transcript_gencode_basic   \n",
       "16                  transcript_appris   \n",
       "17                 external_gene_name   \n",
       "18               external_gene_source   \n",
       "19           external_transcript_name   \n",
       "20    external_transcript_source_name   \n",
       "21                   transcript_count   \n",
       "22         percentage_gene_gc_content   \n",
       "23                       gene_biotype   \n",
       "24                 transcript_biotype   \n",
       "25                             source   \n",
       "26                  transcript_source   \n",
       "27                             status   \n",
       "28                  transcript_status   \n",
       "29                            version   \n",
       "...                               ...   \n",
       "1447                  uniprotsptrembl   \n",
       "1448                cdna_coding_start   \n",
       "1449                  cdna_coding_end   \n",
       "1450                      5_utr_start   \n",
       "1451                        5_utr_end   \n",
       "1452                      3_utr_start   \n",
       "1453                        3_utr_end   \n",
       "1454            ensembl_transcript_id   \n",
       "1455               ensembl_peptide_id   \n",
       "1456               transcript_biotype   \n",
       "1457                           strand   \n",
       "1458                 transcript_start   \n",
       "1459                   transcript_end   \n",
       "1460         transcription_start_site   \n",
       "1461                transcript_length   \n",
       "1462                       cds_length   \n",
       "1463                        cds_start   \n",
       "1464                          cds_end   \n",
       "1465                  ensembl_exon_id   \n",
       "1466                 exon_chrom_start   \n",
       "1467                   exon_chrom_end   \n",
       "1468                           strand   \n",
       "1469                             rank   \n",
       "1470                            phase   \n",
       "1471                        end_phase   \n",
       "1472                cdna_coding_start   \n",
       "1473                  cdna_coding_end   \n",
       "1474             genomic_coding_start   \n",
       "1475               genomic_coding_end   \n",
       "1476                  is_constitutive   \n",
       "\n",
       "                                     description  \n",
       "0                                 Gene stable ID  \n",
       "1                           Transcript stable ID  \n",
       "2                              Protein stable ID  \n",
       "3                                 Exon stable ID  \n",
       "4                               Gene description  \n",
       "5                       Chromosome/scaffold name  \n",
       "6                                Gene Start (bp)  \n",
       "7                                  Gene End (bp)  \n",
       "8                                         Strand  \n",
       "9                                 Karyotype band  \n",
       "10                         Transcript start (bp)  \n",
       "11                           Transcript end (bp)  \n",
       "12                Transcription start site (TSS)  \n",
       "13    Transcript length (including UTRs and CDS)  \n",
       "14                Transcript support level (TSL)  \n",
       "15                      GENCODE basic annotation  \n",
       "16                             APPRIS annotation  \n",
       "17                                     Gene name  \n",
       "18                           Source of gene name  \n",
       "19                               Transcript name  \n",
       "20                     Source of transcript name  \n",
       "21                              Transcript count  \n",
       "22                                  % GC content  \n",
       "23                                     Gene type  \n",
       "24                               Transcript type  \n",
       "25                                 Source (gene)  \n",
       "26                           Source (transcript)  \n",
       "27                                 Status (gene)  \n",
       "28                           Status (transcript)  \n",
       "29                                Version (gene)  \n",
       "...                                          ...  \n",
       "1447                         UniProtKB/TrEMBL ID  \n",
       "1448                     CDS start (within cDNA)  \n",
       "1449                       CDS end (within cDNA)  \n",
       "1450                                5' UTR start  \n",
       "1451                                  5' UTR end  \n",
       "1452                                3' UTR start  \n",
       "1453                                  3' UTR end  \n",
       "1454                        Transcript stable ID  \n",
       "1455                           Protein stable ID  \n",
       "1456                             Transcript type  \n",
       "1457                                      Strand  \n",
       "1458                       Transcript start (bp)  \n",
       "1459                         Transcript end (bp)  \n",
       "1460              Transcription start site (TSS)  \n",
       "1461  Transcript length (including UTRs and CDS)  \n",
       "1462                                  CDS Length  \n",
       "1463                                   CDS start  \n",
       "1464                                     CDS end  \n",
       "1465                              Exon stable ID  \n",
       "1466                      Exon region start (bp)  \n",
       "1467                        Exon region end (bp)  \n",
       "1468                                      Strand  \n",
       "1469                     Exon Rank in Transcript  \n",
       "1470                                 start phase  \n",
       "1471                                   end phase  \n",
       "1472                           cDNA coding start  \n",
       "1473                             cDNA coding end  \n",
       "1474                        Genomic coding start  \n",
       "1475                          Genomic coding end  \n",
       "1476                           Constitutive Exon  \n",
       "\n",
       "[1477 rows x 2 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attributes = bio.list_attributes('ENSEMBL_MART_ENSEMBL', human_dataset)\n",
    "attributes_names = attributes['name'].tolist()\n",
    "attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "data = bio.make_lookup('ENSEMBL_MART_ENSEMBL', human_dataset, attributes=['description'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Z6 small nucleolar RNA [Source:RFAM;Acc:RF00285]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pvt1 oncogene conserved region 3 [Source:RFAM;Acc:RF02166]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U6 spliceosomal RNA [Source:RFAM;Acc:RF00026]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 7SL, cytoplasmic 836, pseudogene [Source:HGNC Symbol;Acc:HGNC:46852]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U2 small nuclear 44, pseudogene [Source:HGNC Symbol;Acc:HGNC:48537]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microRNA 6832 [Source:HGNC Symbol;Acc:HGNC:50140]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y RNA [Source:RFAM;Acc:RF00019]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microRNA 4659b [Source:HGNC Symbol;Acc:HGNC:41530]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Six3os1 conserved region 1 [Source:RFAM;Acc:RF02246]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53 target 1 conserved region 1 [Source:RFAM;Acc:RF02195]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5.8S ribosomal RNA [Source:RFAM;Acc:RF00002]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U1 spliceosomal RNA [Source:RFAM;Acc:RF00003]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vault RNA [Source:RFAM;Acc:RF00006]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U6 small nuclear 673, pseudogene [Source:HGNC Symbol;Acc:HGNC:47636]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 7SL, cytoplasmic 221, pseudogene [Source:HGNC Symbol;Acc:HGNC:46237]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U6 small nuclear 2 [Source:HGNC Symbol;Acc:HGNC:34270]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microRNA 1202 [Source:HGNC Symbol;Acc:HGNC:35268]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microRNA 302c [Source:HGNC Symbol;Acc:HGNC:31764]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 5S ribosomal pseudogene 236 [Source:HGNC Symbol;Acc:HGNC:43136]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microRNA 4435-1 [Source:HGNC Symbol;Acc:HGNC:41713]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KCNQ1 overlapping transcript 1 conserved region 3 [Source:RFAM;Acc:RF01948]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microRNA 4633 [Source:HGNC Symbol;Acc:HGNC:41822]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U6 small nuclear 1179, pseudogene [Source:HGNC Symbol;Acc:HGNC:48142]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microRNA 2054 [Source:HGNC Symbol;Acc:HGNC:37070]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microRNA 6806 [Source:HGNC Symbol;Acc:HGNC:49937]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microRNA 6089 [Source:HGNC Symbol;Acc:HGNC:50179]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 7SL, cytoplasmic 229, pseudogene [Source:HGNC Symbol;Acc:HGNC:46245]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microRNA 4277 [Source:HGNC Symbol;Acc:HGNC:38367]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Small Cajal body specific RNA 20 [Source:RFAM;Acc:RF00601]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Metazoan signal recognition particle RNA [Source:RFAM;Acc:RF00017]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U6 small nuclear 780, pseudogene [Source:HGNC Symbol;Acc:HGNC:47743]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U7 small nuclear 171 pseudogene [Source:HGNC Symbol;Acc:HGNC:45705]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U1 small nuclear 42, pseudogene [Source:HGNC Symbol;Acc:HGNC:48384]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 5S ribosomal pseudogene 251 [Source:HGNC Symbol;Acc:HGNC:43151]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U1 small nuclear 65, pseudogene [Source:HGNC Symbol;Acc:HGNC:48407]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM13A antisense RNA 1 conserved region 2 [Source:RFAM;Acc:RF02115]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U1 small nuclear 27, pseudogene [Source:HGNC Symbol;Acc:HGNC:37497]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U7 small nuclear 75 pseudogene [Source:HGNC Symbol;Acc:HGNC:34171]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 5S ribosomal pseudogene 294 [Source:HGNC Symbol;Acc:HGNC:43194]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U6 small nuclear 466, pseudogene [Source:HGNC Symbol;Acc:HGNC:47429]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 5S ribosomal pseudogene 496 [Source:HGNC Symbol;Acc:HGNC:43396]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U6 small nuclear 553, pseudogene [Source:HGNC Symbol;Acc:HGNC:47516]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>small nucleolar RNA, C/D box 13 [Source:HGNC Symbol;Acc:HGNC:32711]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 7SL, cytoplasmic 185, pseudogene [Source:HGNC Symbol;Acc:HGNC:46201]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U6 small nuclear 1176, pseudogene [Source:HGNC Symbol;Acc:HGNC:48139]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 7SL, cytoplasmic 677, pseudogene [Source:HGNC Symbol;Acc:HGNC:46693]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 5S ribosomal pseudogene 449 [Source:HGNC Symbol;Acc:HGNC:43349]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>small nucleolar RNA, C/D box 42B [Source:HGNC Symbol;Acc:HGNC:10181]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 5S ribosomal pseudogene 338 [Source:HGNC Symbol;Acc:HGNC:43238]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 7SK small nuclear pseudogene 103 [Source:HGNC Symbol;Acc:HGNC:45827]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U6 small nuclear 599, pseudogene [Source:HGNC Symbol;Acc:HGNC:47562]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, 7SL, cytoplasmic 375, pseudogene [Source:HGNC Symbol;Acc:HGNC:46391]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U6 small nuclear 396, pseudogene [Source:HGNC Symbol;Acc:HGNC:47359]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U7 small nuclear 164 pseudogene [Source:HGNC Symbol;Acc:HGNC:45698]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U6 small nuclear 689, pseudogene [Source:HGNC Symbol;Acc:HGNC:47652]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, Ro-associated Y1 pseudogene 13 [Source:HGNC Symbol;Acc:HGNC:50876]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RING1 and YY1 binding protein [Source:EntrezGene;Acc:23429]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B melanoma antigen family member 3 [Source:EntrezGene;Acc:85318]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neuronal acetylcholine receptor subunit alpha-7 isoform 1 precursor  [Source:RefSeq peptide;Acc:NP_000737]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA, U6 small nuclear 443, pseudogene [Source:HGNC Symbol;Acc:HGNC:47406]</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37239 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [Z6 small nucleolar RNA [Source:RFAM;Acc:RF00285], Pvt1 oncogene conserved region 3 [Source:RFAM;Acc:RF02166], U6 spliceosomal RNA [Source:RFAM;Acc:RF00026], RNA, 7SL, cytoplasmic 836, pseudogene [Source:HGNC Symbol;Acc:HGNC:46852], RNA, U2 small nuclear 44, pseudogene [Source:HGNC Symbol;Acc:HGNC:48537], microRNA 6832 [Source:HGNC Symbol;Acc:HGNC:50140], Y RNA [Source:RFAM;Acc:RF00019], microRNA 4659b [Source:HGNC Symbol;Acc:HGNC:41530], Six3os1 conserved region 1 [Source:RFAM;Acc:RF02246], TP53 target 1 conserved region 1 [Source:RFAM;Acc:RF02195], 5.8S ribosomal RNA [Source:RFAM;Acc:RF00002], U1 spliceosomal RNA [Source:RFAM;Acc:RF00003], Vault RNA [Source:RFAM;Acc:RF00006], RNA, U6 small nuclear 673, pseudogene [Source:HGNC Symbol;Acc:HGNC:47636], RNA, 7SL, cytoplasmic 221, pseudogene [Source:HGNC Symbol;Acc:HGNC:46237], RNA, U6 small nuclear 2 [Source:HGNC Symbol;Acc:HGNC:34270], microRNA 1202 [Source:HGNC Symbol;Acc:HGNC:35268], microRNA 302c [Source:HGNC Symbol;Acc:HGNC:31764], RNA, 5S ribosomal pseudogene 236 [Source:HGNC Symbol;Acc:HGNC:43136], microRNA 4435-1 [Source:HGNC Symbol;Acc:HGNC:41713], KCNQ1 overlapping transcript 1 conserved region 3 [Source:RFAM;Acc:RF01948], microRNA 4633 [Source:HGNC Symbol;Acc:HGNC:41822], RNA, U6 small nuclear 1179, pseudogene [Source:HGNC Symbol;Acc:HGNC:48142], microRNA 2054 [Source:HGNC Symbol;Acc:HGNC:37070], microRNA 6806 [Source:HGNC Symbol;Acc:HGNC:49937], microRNA 6089 [Source:HGNC Symbol;Acc:HGNC:50179], RNA, 7SL, cytoplasmic 229, pseudogene [Source:HGNC Symbol;Acc:HGNC:46245], microRNA 4277 [Source:HGNC Symbol;Acc:HGNC:38367], Small Cajal body specific RNA 20 [Source:RFAM;Acc:RF00601], Metazoan signal recognition particle RNA [Source:RFAM;Acc:RF00017], RNA, U2 small nuclear 64, pseudogene [Source:HGNC Symbol;Acc:HGNC:48557], RNA, 5S ribosomal pseudogene 508 [Source:HGNC Symbol;Acc:HGNC:43408], microRNA 1281 [Source:HGNC Symbol;Acc:HGNC:35359], microRNA 1283-1 [Source:HGNC Symbol;Acc:HGNC:35255], RNA, U1 small nuclear 7, pseudogene [Source:HGNC Symbol;Acc:HGNC:48306], RNA, U2 small nuclear 13, pseudogene [Source:HGNC Symbol;Acc:HGNC:48506], microRNA 7151 [Source:HGNC Symbol;Acc:HGNC:50170], RNA, 5S ribosomal pseudogene 94 [Source:HGNC Symbol;Acc:HGNC:42892], microRNA 4293 [Source:HGNC Symbol;Acc:HGNC:38270], RNA, U7 small nuclear 67 pseudogene [Source:HGNC Symbol;Acc:HGNC:34163], RNA, U6 small nuclear 572, pseudogene [Source:HGNC Symbol;Acc:HGNC:47535], T-cell-interacting, activating receptor on myeloid cells 1 [Source:HGNC Symbol;Acc:HGNC:37250], U8 small nucleolar RNA [Source:RFAM;Acc:RF00096], RNA, U5E small nuclear 6, pseudogene [Source:HGNC Symbol;Acc:HGNC:42522], microRNA 3137 [Source:HGNC Symbol;Acc:HGNC:38263], Taurine upregulated gene 1 conserved region 2 [Source:RFAM;Acc:RF01883], microRNA 548ae-1 [Source:HGNC Symbol;Acc:HGNC:41742], small nucleolar RNA, H/ACA box 30 [Source:HGNC Symbol;Acc:HGNC:32620], RNA, 5S ribosomal pseudogene 37 [Source:HGNC Symbol;Acc:HGNC:42613], microRNA 4273 [Source:HGNC Symbol;Acc:HGNC:38339], RNA, 5S ribosomal pseudogene 107 [Source:HGNC Symbol;Acc:HGNC:42905], microRNA 154 [Source:HGNC Symbol;Acc:HGNC:31541], RNA, U6 small nuclear 519, pseudogene [Source:HGNC Symbol;Acc:HGNC:47482], RNA, U6 small nuclear 10, pseudogene [Source:HGNC Symbol;Acc:HGNC:34254], small nucleolar RNA, C/D box 92 [Source:HGNC Symbol;Acc:HGNC:32754], RNA, 7SK small nuclear pseudogene 178 [Source:HGNC Symbol;Acc:HGNC:45902], Small nucleolar RNA SNORD46 [Source:RFAM;Acc:RF00218], RNA, 7SL, cytoplasmic 123, pseudogene [Source:HGNC Symbol;Acc:HGNC:46139], microRNA 3917 [Source:HGNC Symbol;Acc:HGNC:38911], microRNA 1183 [Source:HGNC Symbol;Acc:HGNC:35264], RNA, U5B small nuclear 4, pseudogene [Source:HGNC Symbol;Acc:HGNC:34184], TCF3 fusion partner [Source:HGNC Symbol;Acc:HGNC:13630], microRNA 5696 [Source:HGNC Symbol;Acc:HGNC:43516], microRNA 3175 [Source:HGNC Symbol;Acc:HGNC:38299], Small Cajal body specific RNA 11 [Source:RFAM;Acc:RF00564], microRNA 3913-2 [Source:HGNC Symbol;Acc:HGNC:38953], DLG2 antisense RNA 1 conserved region 1 [Source:RFAM;Acc:RF02112], RNA, 7SK small nuclear pseudogene 293 [Source:HGNC Symbol;Acc:HGNC:46017], JPX transcript, XIST activator conserved region 1 [Source:RFAM;Acc:RF02124], RNA, U5A small nuclear 6, pseudogene [Source:HGNC Symbol;Acc:HGNC:42527], small nucleolar RNA, C/D box 32B [Source:HGNC Symbol;Acc:HGNC:32719], X-chromosome inactivation gene exon 1 [Source:RFAM;Acc:RF01880], TUC338 [Source:RFAM;Acc:RF02271], RNA, 5S ribosomal pseudogene 109 [Source:HGNC Symbol;Acc:HGNC:42907], 5S ribosomal RNA [Source:RFAM;Acc:RF00001], Small nucleolar RNA SNORA1 [Source:RFAM;Acc:RF00408], Small nucleolar RNA SNORA43 [Source:RFAM;Acc:RF00416], Small Cajal body specific RNA 21 [Source:RFAM;Acc:RF00602], transcription elongation factor A like 9 [Source:HGNC Symbol;Acc:HGNC:30084], RNA, 7SL, cytoplasmic 741, pseudogene [Source:HGNC Symbol;Acc:HGNC:46757], RNA, U6 small nuclear 107, pseudogene [Source:HGNC Symbol;Acc:HGNC:47070], RNA, 7SL, cytoplasmic 659, pseudogene [Source:HGNC Symbol;Acc:HGNC:46675], Small nucleolar RNA SNORD30 [Source:RFAM;Acc:RF00088], RNA, U4 small nuclear 91, pseudogene [Source:HGNC Symbol;Acc:HGNC:47027], RNA, 7SL, cytoplasmic 278, pseudogene [Source:HGNC Symbol;Acc:HGNC:46294], microRNA 8052 [Source:HGNC Symbol;Acc:HGNC:50252], small nucleolar RNA, H/ACA box 38 [Source:HGNC Symbol;Acc:HGNC:32631], Small nucleolar RNA Me28S-Am2634 [Source:RFAM;Acc:RF00537], small nucleolar RNA, H/ACA box 16B [Source:HGNC Symbol;Acc:HGNC:32606], RNA, 7SL, cytoplasmic 107, pseudogene [Source:HGNC Symbol;Acc:HGNC:46123], RNA, U6 small nuclear 657, pseudogene [Source:HGNC Symbol;Acc:HGNC:47620], RNA, U6 small nuclear 737, pseudogene [Source:HGNC Symbol;Acc:HGNC:47700], RNA, U7 small nuclear 28 pseudogene [Source:HGNC Symbol;Acc:HGNC:34124], RNA, U4 small nuclear 63, pseudogene [Source:HGNC Symbol;Acc:HGNC:46999], microRNA 8089 [Source:HGNC Symbol;Acc:HGNC:50196], RNA, 7SK small nuclear pseudogene 117 [Source:HGNC Symbol;Acc:HGNC:45841], microRNA 5739 [Source:HGNC Symbol;Acc:HGNC:49977], Small nucleolar RNA SNORA40 [Source:RFAM;Acc:RF00561], microRNA 519b [Source:HGNC Symbol;Acc:HGNC:32101], RNA, 7SL, cytoplasmic 160, pseudogene [Source:HGNC Symbol;Acc:HGNC:46176], ...]\n",
       "\n",
       "[37239 rows x 0 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
