{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: biomaRt\n"
     ]
    }
   ],
   "source": [
    "require(biomaRt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>biomart</th><th scope=col>version</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>ENSEMBL_MART_ENSEMBL </td><td>Ensembl Genes 88     </td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_MOUSE   </td><td>Mouse strains 88     </td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_SNP     </td><td>Ensembl Variation 88 </td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_FUNCGEN </td><td>Ensembl Regulation 88</td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_VEGA    </td><td>Vega 68              </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       " biomart & version\\\\\n",
       "\\hline\n",
       "\t ENSEMBL\\_MART\\_ENSEMBL & Ensembl Genes 88        \\\\\n",
       "\t ENSEMBL\\_MART\\_MOUSE   & Mouse strains 88        \\\\\n",
       "\t ENSEMBL\\_MART\\_SNP     & Ensembl Variation 88    \\\\\n",
       "\t ENSEMBL\\_MART\\_FUNCGEN & Ensembl Regulation 88   \\\\\n",
       "\t ENSEMBL\\_MART\\_VEGA    & Vega 68                 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "biomart | version | \n",
       "|---|---|---|---|---|\n",
       "| ENSEMBL_MART_ENSEMBL  | Ensembl Genes 88      | \n",
       "| ENSEMBL_MART_MOUSE    | Mouse strains 88      | \n",
       "| ENSEMBL_MART_SNP      | Ensembl Variation 88  | \n",
       "| ENSEMBL_MART_FUNCGEN  | Ensembl Regulation 88 | \n",
       "| ENSEMBL_MART_VEGA     | Vega 68               | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  biomart              version              \n",
       "1 ENSEMBL_MART_ENSEMBL Ensembl Genes 88     \n",
       "2 ENSEMBL_MART_MOUSE   Mouse strains 88     \n",
       "3 ENSEMBL_MART_SNP     Ensembl Variation 88 \n",
       "4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 88\n",
       "5 ENSEMBL_MART_VEGA    Vega 68              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listMarts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ensembl=useMart(\"ensembl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>dataset</th><th scope=col>description</th><th scope=col>version</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>oanatinus_gene_ensembl                     </td><td>Platypus genes (OANA5)                     </td><td>OANA5                                      </td></tr>\n",
       "\t<tr><td>cporcellus_gene_ensembl                    </td><td>Guinea Pig genes (cavPor3)                 </td><td>cavPor3                                    </td></tr>\n",
       "\t<tr><td>gaculeatus_gene_ensembl                    </td><td>Stickleback genes (BROAD S1)               </td><td>BROAD S1                                   </td></tr>\n",
       "\t<tr><td>lafricana_gene_ensembl                     </td><td>Elephant genes (Loxafr3.0)                 </td><td>Loxafr3.0                                  </td></tr>\n",
       "\t<tr><td>itridecemlineatus_gene_ensembl             </td><td>Squirrel genes (spetri2)                   </td><td>spetri2                                    </td></tr>\n",
       "\t<tr><td>choffmanni_gene_ensembl                    </td><td>Sloth genes (choHof1)                      </td><td>choHof1                                    </td></tr>\n",
       "\t<tr><td>fcatus_gene_ensembl                        </td><td>Cat genes (Felis_catus_6.2)                </td><td>Felis_catus_6.2                            </td></tr>\n",
       "\t<tr><td>csavignyi_gene_ensembl                     </td><td>C.savignyi genes (CSAV 2.0)                </td><td>CSAV 2.0                                   </td></tr>\n",
       "\t<tr><td>rnorvegicus_gene_ensembl                   </td><td>Rat genes (Rnor_6.0)                       </td><td>Rnor_6.0                                   </td></tr>\n",
       "\t<tr><td>psinensis_gene_ensembl                     </td><td>Chinese softshell turtle genes (PelSin_1.0)</td><td>PelSin_1.0                                 </td></tr>\n",
       "\t<tr><td>cjacchus_gene_ensembl                      </td><td>Marmoset genes (C_jacchus3.2.1)            </td><td>C_jacchus3.2.1                             </td></tr>\n",
       "\t<tr><td>ttruncatus_gene_ensembl                    </td><td>Dolphin genes (turTru1)                    </td><td>turTru1                                    </td></tr>\n",
       "\t<tr><td>scerevisiae_gene_ensembl                   </td><td>Saccharomyces cerevisiae genes (R64-1-1)   </td><td>R64-1-1                                    </td></tr>\n",
       "\t<tr><td>celegans_gene_ensembl                      </td><td>Caenorhabditis elegans genes (WBcel235)    </td><td>WBcel235                                   </td></tr>\n",
       "\t<tr><td>csabaeus_gene_ensembl                      </td><td>Vervet-AGM genes (ChlSab1.1)               </td><td>ChlSab1.1                                  </td></tr>\n",
       "\t<tr><td>oniloticus_gene_ensembl                    </td><td>Tilapia genes (Orenil1.0)                  </td><td>Orenil1.0                                  </td></tr>\n",
       "\t<tr><td>trubripes_gene_ensembl                     </td><td>Fugu genes (FUGU 4.0)                      </td><td>FUGU 4.0                                   </td></tr>\n",
       "\t<tr><td>amexicanus_gene_ensembl                    </td><td>Cave fish genes (AstMex102)                </td><td>AstMex102                                  </td></tr>\n",
       "\t<tr><td>pmarinus_gene_ensembl                      </td><td>Lamprey genes (Pmarinus_7.0)               </td><td>Pmarinus_7.0                               </td></tr>\n",
       "\t<tr><td>eeuropaeus_gene_ensembl                    </td><td>Hedgehog genes (eriEur1)                   </td><td>eriEur1                                    </td></tr>\n",
       "\t<tr><td>falbicollis_gene_ensembl                   </td><td>Flycatcher genes (FicAlb_1.4)              </td><td>FicAlb_1.4                                 </td></tr>\n",
       "\t<tr><td>ptroglodytes_gene_ensembl                  </td><td>Chimpanzee genes (CHIMP2.1.4)              </td><td>CHIMP2.1.4                                 </td></tr>\n",
       "\t<tr><td>cintestinalis_gene_ensembl                 </td><td>C.intestinalis genes (KH)                  </td><td>KH                                         </td></tr>\n",
       "\t<tr><td>etelfairi_gene_ensembl                     </td><td>Lesser hedgehog tenrec genes (TENREC)      </td><td>TENREC                                     </td></tr>\n",
       "\t<tr><td>nleucogenys_gene_ensembl                   </td><td>Gibbon genes (Nleu1.0)                     </td><td>Nleu1.0                                    </td></tr>\n",
       "\t<tr><td>sscrofa_gene_ensembl                       </td><td>Pig genes (Sscrofa10.2)                    </td><td>Sscrofa10.2                                </td></tr>\n",
       "\t<tr><td>ocuniculus_gene_ensembl                    </td><td>Rabbit genes (OryCun2.0)                   </td><td>OryCun2.0                                  </td></tr>\n",
       "\t<tr><td>dnovemcinctus_gene_ensembl                 </td><td>Armadillo genes (Dasnov3.0)                </td><td>Dasnov3.0                                  </td></tr>\n",
       "\t<tr><td>pcapensis_gene_ensembl                     </td><td>Hyrax genes (proCap1)                      </td><td>proCap1                                    </td></tr>\n",
       "\t<tr><td>tguttata_gene_ensembl                      </td><td>Zebra Finch genes (taeGut3.2.4)            </td><td>taeGut3.2.4                                </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>xmaculatus_gene_ensembl               </td><td>Platyfish genes (Xipmac4.4.2)         </td><td>Xipmac4.4.2                           </td></tr>\n",
       "\t<tr><td>drerio_gene_ensembl                   </td><td>Zebrafish genes (GRCz10)              </td><td>GRCz10                                </td></tr>\n",
       "\t<tr><td>tnigroviridis_gene_ensembl            </td><td>Tetraodon genes (TETRAODON 8.0)       </td><td>TETRAODON 8.0                         </td></tr>\n",
       "\t<tr><td>lchalumnae_gene_ensembl               </td><td>Coelacanth genes (LatCha1)            </td><td>LatCha1                               </td></tr>\n",
       "\t<tr><td>amelanoleuca_gene_ensembl             </td><td>Panda genes (ailMel1)                 </td><td>ailMel1                               </td></tr>\n",
       "\t<tr><td>mmulatta_gene_ensembl                 </td><td>Macaque genes (Mmul_8.0.1)            </td><td>Mmul_8.0.1                            </td></tr>\n",
       "\t<tr><td>pvampyrus_gene_ensembl                </td><td>Megabat genes (pteVam1)               </td><td>pteVam1                               </td></tr>\n",
       "\t<tr><td>panubis_gene_ensembl                  </td><td>Olive baboon genes (PapAnu2.0)        </td><td>PapAnu2.0                             </td></tr>\n",
       "\t<tr><td>mdomestica_gene_ensembl               </td><td>Opossum genes (monDom5)               </td><td>monDom5                               </td></tr>\n",
       "\t<tr><td>acarolinensis_gene_ensembl            </td><td>Anole lizard genes (AnoCar2.0)        </td><td>AnoCar2.0                             </td></tr>\n",
       "\t<tr><td>vpacos_gene_ensembl                   </td><td>Alpaca genes (vicPac1)                </td><td>vicPac1                               </td></tr>\n",
       "\t<tr><td>tsyrichta_gene_ensembl                </td><td>Tarsier genes (tarSyr1)               </td><td>tarSyr1                               </td></tr>\n",
       "\t<tr><td>ogarnettii_gene_ensembl               </td><td>Bushbaby genes (OtoGar3)              </td><td>OtoGar3                               </td></tr>\n",
       "\t<tr><td>dmelanogaster_gene_ensembl            </td><td>Fruitfly genes (BDGP6)                </td><td>BDGP6                                 </td></tr>\n",
       "\t<tr><td>loculatus_gene_ensembl                </td><td>Spotted gar genes (LepOcu1)           </td><td>LepOcu1                               </td></tr>\n",
       "\t<tr><td>mmurinus_gene_ensembl                 </td><td>Mouse Lemur genes (Mmur_2.0)          </td><td>Mmur_2.0                              </td></tr>\n",
       "\t<tr><td>olatipes_gene_ensembl                 </td><td>Medaka genes (HdrR)                   </td><td>HdrR                                  </td></tr>\n",
       "\t<tr><td>oprinceps_gene_ensembl                </td><td>Pika genes (OchPri2.0-Ens)            </td><td>OchPri2.0-Ens                         </td></tr>\n",
       "\t<tr><td>ggorilla_gene_ensembl                 </td><td>Gorilla genes (gorGor3.1)             </td><td>gorGor3.1                             </td></tr>\n",
       "\t<tr><td>dordii_gene_ensembl                   </td><td>Kangaroo rat genes (dipOrd1)          </td><td>dipOrd1                               </td></tr>\n",
       "\t<tr><td>oaries_gene_ensembl                   </td><td>Sheep genes (Oar_v3.1)                </td><td>Oar_v3.1                              </td></tr>\n",
       "\t<tr><td>mmusculus_gene_ensembl                </td><td>Mouse genes (GRCm38.p5)               </td><td>GRCm38.p5                             </td></tr>\n",
       "\t<tr><td>mgallopavo_gene_ensembl               </td><td>Turkey genes (Turkey_2.01)            </td><td>Turkey_2.01                           </td></tr>\n",
       "\t<tr><td>gmorhua_gene_ensembl                  </td><td>Cod genes (gadMor1)                   </td><td>gadMor1                               </td></tr>\n",
       "\t<tr><td>saraneus_gene_ensembl                 </td><td>Shrew genes (sorAra1)                 </td><td>sorAra1                               </td></tr>\n",
       "\t<tr><td>aplatyrhynchos_gene_ensembl           </td><td>Duck genes (BGI_duck_1.0)             </td><td>BGI_duck_1.0                          </td></tr>\n",
       "\t<tr><td>sharrisii_gene_ensembl                </td><td>Tasmanian devil genes (Devil_ref v7.0)</td><td>Devil_ref v7.0                        </td></tr>\n",
       "\t<tr><td>btaurus_gene_ensembl                  </td><td>Cow genes (UMD3.1)                    </td><td>UMD3.1                                </td></tr>\n",
       "\t<tr><td>meugenii_gene_ensembl                 </td><td>Wallaby genes (Meug_1.0)              </td><td>Meug_1.0                              </td></tr>\n",
       "\t<tr><td>cfamiliaris_gene_ensembl              </td><td>Dog genes (CanFam3.1)                 </td><td>CanFam3.1                             </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lll}\n",
       " dataset & description & version\\\\\n",
       "\\hline\n",
       "\t oanatinus\\_gene\\_ensembl                  & Platypus genes (OANA5)                      & OANA5                                      \\\\\n",
       "\t cporcellus\\_gene\\_ensembl                 & Guinea Pig genes (cavPor3)                  & cavPor3                                    \\\\\n",
       "\t gaculeatus\\_gene\\_ensembl                 & Stickleback genes (BROAD S1)                & BROAD S1                                   \\\\\n",
       "\t lafricana\\_gene\\_ensembl                  & Elephant genes (Loxafr3.0)                  & Loxafr3.0                                  \\\\\n",
       "\t itridecemlineatus\\_gene\\_ensembl          & Squirrel genes (spetri2)                    & spetri2                                    \\\\\n",
       "\t choffmanni\\_gene\\_ensembl                 & Sloth genes (choHof1)                       & choHof1                                    \\\\\n",
       "\t fcatus\\_gene\\_ensembl                         & Cat genes (Felis\\_catus\\_6.2)                 & Felis\\_catus\\_6.2                            \\\\\n",
       "\t csavignyi\\_gene\\_ensembl                  & C.savignyi genes (CSAV 2.0)                 & CSAV 2.0                                   \\\\\n",
       "\t rnorvegicus\\_gene\\_ensembl                  & Rat genes (Rnor\\_6.0)                        & Rnor\\_6.0                                   \\\\\n",
       "\t psinensis\\_gene\\_ensembl                    & Chinese softshell turtle genes (PelSin\\_1.0) & PelSin\\_1.0                                 \\\\\n",
       "\t cjacchus\\_gene\\_ensembl                     & Marmoset genes (C\\_jacchus3.2.1)             & C\\_jacchus3.2.1                             \\\\\n",
       "\t ttruncatus\\_gene\\_ensembl                 & Dolphin genes (turTru1)                     & turTru1                                    \\\\\n",
       "\t scerevisiae\\_gene\\_ensembl                & Saccharomyces cerevisiae genes (R64-1-1)    & R64-1-1                                    \\\\\n",
       "\t celegans\\_gene\\_ensembl                   & Caenorhabditis elegans genes (WBcel235)     & WBcel235                                   \\\\\n",
       "\t csabaeus\\_gene\\_ensembl                   & Vervet-AGM genes (ChlSab1.1)                & ChlSab1.1                                  \\\\\n",
       "\t oniloticus\\_gene\\_ensembl                 & Tilapia genes (Orenil1.0)                   & Orenil1.0                                  \\\\\n",
       "\t trubripes\\_gene\\_ensembl                  & Fugu genes (FUGU 4.0)                       & FUGU 4.0                                   \\\\\n",
       "\t amexicanus\\_gene\\_ensembl                 & Cave fish genes (AstMex102)                 & AstMex102                                  \\\\\n",
       "\t pmarinus\\_gene\\_ensembl                     & Lamprey genes (Pmarinus\\_7.0)                & Pmarinus\\_7.0                               \\\\\n",
       "\t eeuropaeus\\_gene\\_ensembl                 & Hedgehog genes (eriEur1)                    & eriEur1                                    \\\\\n",
       "\t falbicollis\\_gene\\_ensembl                  & Flycatcher genes (FicAlb\\_1.4)               & FicAlb\\_1.4                                 \\\\\n",
       "\t ptroglodytes\\_gene\\_ensembl               & Chimpanzee genes (CHIMP2.1.4)               & CHIMP2.1.4                                 \\\\\n",
       "\t cintestinalis\\_gene\\_ensembl              & C.intestinalis genes (KH)                   & KH                                         \\\\\n",
       "\t etelfairi\\_gene\\_ensembl                  & Lesser hedgehog tenrec genes (TENREC)       & TENREC                                     \\\\\n",
       "\t nleucogenys\\_gene\\_ensembl                & Gibbon genes (Nleu1.0)                      & Nleu1.0                                    \\\\\n",
       "\t sscrofa\\_gene\\_ensembl                    & Pig genes (Sscrofa10.2)                     & Sscrofa10.2                                \\\\\n",
       "\t ocuniculus\\_gene\\_ensembl                 & Rabbit genes (OryCun2.0)                    & OryCun2.0                                  \\\\\n",
       "\t dnovemcinctus\\_gene\\_ensembl              & Armadillo genes (Dasnov3.0)                 & Dasnov3.0                                  \\\\\n",
       "\t pcapensis\\_gene\\_ensembl                  & Hyrax genes (proCap1)                       & proCap1                                    \\\\\n",
       "\t tguttata\\_gene\\_ensembl                   & Zebra Finch genes (taeGut3.2.4)             & taeGut3.2.4                                \\\\\n",
       "\t ⋮ & ⋮ & ⋮\\\\\n",
       "\t xmaculatus\\_gene\\_ensembl            & Platyfish genes (Xipmac4.4.2)          & Xipmac4.4.2                           \\\\\n",
       "\t drerio\\_gene\\_ensembl                & Zebrafish genes (GRCz10)               & GRCz10                                \\\\\n",
       "\t tnigroviridis\\_gene\\_ensembl         & Tetraodon genes (TETRAODON 8.0)        & TETRAODON 8.0                         \\\\\n",
       "\t lchalumnae\\_gene\\_ensembl            & Coelacanth genes (LatCha1)             & LatCha1                               \\\\\n",
       "\t amelanoleuca\\_gene\\_ensembl          & Panda genes (ailMel1)                  & ailMel1                               \\\\\n",
       "\t mmulatta\\_gene\\_ensembl                & Macaque genes (Mmul\\_8.0.1)             & Mmul\\_8.0.1                            \\\\\n",
       "\t pvampyrus\\_gene\\_ensembl             & Megabat genes (pteVam1)                & pteVam1                               \\\\\n",
       "\t panubis\\_gene\\_ensembl               & Olive baboon genes (PapAnu2.0)         & PapAnu2.0                             \\\\\n",
       "\t mdomestica\\_gene\\_ensembl            & Opossum genes (monDom5)                & monDom5                               \\\\\n",
       "\t acarolinensis\\_gene\\_ensembl         & Anole lizard genes (AnoCar2.0)         & AnoCar2.0                             \\\\\n",
       "\t vpacos\\_gene\\_ensembl                & Alpaca genes (vicPac1)                 & vicPac1                               \\\\\n",
       "\t tsyrichta\\_gene\\_ensembl             & Tarsier genes (tarSyr1)                & tarSyr1                               \\\\\n",
       "\t ogarnettii\\_gene\\_ensembl            & Bushbaby genes (OtoGar3)               & OtoGar3                               \\\\\n",
       "\t dmelanogaster\\_gene\\_ensembl         & Fruitfly genes (BDGP6)                 & BDGP6                                 \\\\\n",
       "\t loculatus\\_gene\\_ensembl             & Spotted gar genes (LepOcu1)            & LepOcu1                               \\\\\n",
       "\t mmurinus\\_gene\\_ensembl                & Mouse Lemur genes (Mmur\\_2.0)           & Mmur\\_2.0                              \\\\\n",
       "\t olatipes\\_gene\\_ensembl              & Medaka genes (HdrR)                    & HdrR                                  \\\\\n",
       "\t oprinceps\\_gene\\_ensembl             & Pika genes (OchPri2.0-Ens)             & OchPri2.0-Ens                         \\\\\n",
       "\t ggorilla\\_gene\\_ensembl              & Gorilla genes (gorGor3.1)              & gorGor3.1                             \\\\\n",
       "\t dordii\\_gene\\_ensembl                & Kangaroo rat genes (dipOrd1)           & dipOrd1                               \\\\\n",
       "\t oaries\\_gene\\_ensembl                  & Sheep genes (Oar\\_v3.1)                 & Oar\\_v3.1                              \\\\\n",
       "\t mmusculus\\_gene\\_ensembl             & Mouse genes (GRCm38.p5)                & GRCm38.p5                             \\\\\n",
       "\t mgallopavo\\_gene\\_ensembl              & Turkey genes (Turkey\\_2.01)             & Turkey\\_2.01                           \\\\\n",
       "\t gmorhua\\_gene\\_ensembl               & Cod genes (gadMor1)                    & gadMor1                               \\\\\n",
       "\t saraneus\\_gene\\_ensembl              & Shrew genes (sorAra1)                  & sorAra1                               \\\\\n",
       "\t aplatyrhynchos\\_gene\\_ensembl            & Duck genes (BGI\\_duck\\_1.0)              & BGI\\_duck\\_1.0                          \\\\\n",
       "\t sharrisii\\_gene\\_ensembl               & Tasmanian devil genes (Devil\\_ref v7.0) & Devil\\_ref v7.0                        \\\\\n",
       "\t btaurus\\_gene\\_ensembl               & Cow genes (UMD3.1)                     & UMD3.1                                \\\\\n",
       "\t meugenii\\_gene\\_ensembl                & Wallaby genes (Meug\\_1.0)               & Meug\\_1.0                              \\\\\n",
       "\t cfamiliaris\\_gene\\_ensembl           & Dog genes (CanFam3.1)                  & CanFam3.1                             \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "dataset | description | version | \n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| oanatinus_gene_ensembl                      | Platypus genes (OANA5)                      | OANA5                                       | \n",
       "| cporcellus_gene_ensembl                     | Guinea Pig genes (cavPor3)                  | cavPor3                                     | \n",
       "| gaculeatus_gene_ensembl                     | Stickleback genes (BROAD S1)                | BROAD S1                                    | \n",
       "| lafricana_gene_ensembl                      | Elephant genes (Loxafr3.0)                  | Loxafr3.0                                   | \n",
       "| itridecemlineatus_gene_ensembl              | Squirrel genes (spetri2)                    | spetri2                                     | \n",
       "| choffmanni_gene_ensembl                     | Sloth genes (choHof1)                       | choHof1                                     | \n",
       "| fcatus_gene_ensembl                         | Cat genes (Felis_catus_6.2)                 | Felis_catus_6.2                             | \n",
       "| csavignyi_gene_ensembl                      | C.savignyi genes (CSAV 2.0)                 | CSAV 2.0                                    | \n",
       "| rnorvegicus_gene_ensembl                    | Rat genes (Rnor_6.0)                        | Rnor_6.0                                    | \n",
       "| psinensis_gene_ensembl                      | Chinese softshell turtle genes (PelSin_1.0) | PelSin_1.0                                  | \n",
       "| cjacchus_gene_ensembl                       | Marmoset genes (C_jacchus3.2.1)             | C_jacchus3.2.1                              | \n",
       "| ttruncatus_gene_ensembl                     | Dolphin genes (turTru1)                     | turTru1                                     | \n",
       "| scerevisiae_gene_ensembl                    | Saccharomyces cerevisiae genes (R64-1-1)    | R64-1-1                                     | \n",
       "| celegans_gene_ensembl                       | Caenorhabditis elegans genes (WBcel235)     | WBcel235                                    | \n",
       "| csabaeus_gene_ensembl                       | Vervet-AGM genes (ChlSab1.1)                | ChlSab1.1                                   | \n",
       "| oniloticus_gene_ensembl                     | Tilapia genes (Orenil1.0)                   | Orenil1.0                                   | \n",
       "| trubripes_gene_ensembl                      | Fugu genes (FUGU 4.0)                       | FUGU 4.0                                    | \n",
       "| amexicanus_gene_ensembl                     | Cave fish genes (AstMex102)                 | AstMex102                                   | \n",
       "| pmarinus_gene_ensembl                       | Lamprey genes (Pmarinus_7.0)                | Pmarinus_7.0                                | \n",
       "| eeuropaeus_gene_ensembl                     | Hedgehog genes (eriEur1)                    | eriEur1                                     | \n",
       "| falbicollis_gene_ensembl                    | Flycatcher genes (FicAlb_1.4)               | FicAlb_1.4                                  | \n",
       "| ptroglodytes_gene_ensembl                   | Chimpanzee genes (CHIMP2.1.4)               | CHIMP2.1.4                                  | \n",
       "| cintestinalis_gene_ensembl                  | C.intestinalis genes (KH)                   | KH                                          | \n",
       "| etelfairi_gene_ensembl                      | Lesser hedgehog tenrec genes (TENREC)       | TENREC                                      | \n",
       "| nleucogenys_gene_ensembl                    | Gibbon genes (Nleu1.0)                      | Nleu1.0                                     | \n",
       "| sscrofa_gene_ensembl                        | Pig genes (Sscrofa10.2)                     | Sscrofa10.2                                 | \n",
       "| ocuniculus_gene_ensembl                     | Rabbit genes (OryCun2.0)                    | OryCun2.0                                   | \n",
       "| dnovemcinctus_gene_ensembl                  | Armadillo genes (Dasnov3.0)                 | Dasnov3.0                                   | \n",
       "| pcapensis_gene_ensembl                      | Hyrax genes (proCap1)                       | proCap1                                     | \n",
       "| tguttata_gene_ensembl                       | Zebra Finch genes (taeGut3.2.4)             | taeGut3.2.4                                 | \n",
       "| ⋮ | ⋮ | ⋮ | \n",
       "| xmaculatus_gene_ensembl                | Platyfish genes (Xipmac4.4.2)          | Xipmac4.4.2                            | \n",
       "| drerio_gene_ensembl                    | Zebrafish genes (GRCz10)               | GRCz10                                 | \n",
       "| tnigroviridis_gene_ensembl             | Tetraodon genes (TETRAODON 8.0)        | TETRAODON 8.0                          | \n",
       "| lchalumnae_gene_ensembl                | Coelacanth genes (LatCha1)             | LatCha1                                | \n",
       "| amelanoleuca_gene_ensembl              | Panda genes (ailMel1)                  | ailMel1                                | \n",
       "| mmulatta_gene_ensembl                  | Macaque genes (Mmul_8.0.1)             | Mmul_8.0.1                             | \n",
       "| pvampyrus_gene_ensembl                 | Megabat genes (pteVam1)                | pteVam1                                | \n",
       "| panubis_gene_ensembl                   | Olive baboon genes (PapAnu2.0)         | PapAnu2.0                              | \n",
       "| mdomestica_gene_ensembl                | Opossum genes (monDom5)                | monDom5                                | \n",
       "| acarolinensis_gene_ensembl             | Anole lizard genes (AnoCar2.0)         | AnoCar2.0                              | \n",
       "| vpacos_gene_ensembl                    | Alpaca genes (vicPac1)                 | vicPac1                                | \n",
       "| tsyrichta_gene_ensembl                 | Tarsier genes (tarSyr1)                | tarSyr1                                | \n",
       "| ogarnettii_gene_ensembl                | Bushbaby genes (OtoGar3)               | OtoGar3                                | \n",
       "| dmelanogaster_gene_ensembl             | Fruitfly genes (BDGP6)                 | BDGP6                                  | \n",
       "| loculatus_gene_ensembl                 | Spotted gar genes (LepOcu1)            | LepOcu1                                | \n",
       "| mmurinus_gene_ensembl                  | Mouse Lemur genes (Mmur_2.0)           | Mmur_2.0                               | \n",
       "| olatipes_gene_ensembl                  | Medaka genes (HdrR)                    | HdrR                                   | \n",
       "| oprinceps_gene_ensembl                 | Pika genes (OchPri2.0-Ens)             | OchPri2.0-Ens                          | \n",
       "| ggorilla_gene_ensembl                  | Gorilla genes (gorGor3.1)              | gorGor3.1                              | \n",
       "| dordii_gene_ensembl                    | Kangaroo rat genes (dipOrd1)           | dipOrd1                                | \n",
       "| oaries_gene_ensembl                    | Sheep genes (Oar_v3.1)                 | Oar_v3.1                               | \n",
       "| mmusculus_gene_ensembl                 | Mouse genes (GRCm38.p5)                | GRCm38.p5                              | \n",
       "| mgallopavo_gene_ensembl                | Turkey genes (Turkey_2.01)             | Turkey_2.01                            | \n",
       "| gmorhua_gene_ensembl                   | Cod genes (gadMor1)                    | gadMor1                                | \n",
       "| saraneus_gene_ensembl                  | Shrew genes (sorAra1)                  | sorAra1                                | \n",
       "| aplatyrhynchos_gene_ensembl            | Duck genes (BGI_duck_1.0)              | BGI_duck_1.0                           | \n",
       "| sharrisii_gene_ensembl                 | Tasmanian devil genes (Devil_ref v7.0) | Devil_ref v7.0                         | \n",
       "| btaurus_gene_ensembl                   | Cow genes (UMD3.1)                     | UMD3.1                                 | \n",
       "| meugenii_gene_ensembl                  | Wallaby genes (Meug_1.0)               | Meug_1.0                               | \n",
       "| cfamiliaris_gene_ensembl               | Dog genes (CanFam3.1)                  | CanFam3.1                              | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "   dataset                        description                                \n",
       "1  oanatinus_gene_ensembl         Platypus genes (OANA5)                     \n",
       "2  cporcellus_gene_ensembl        Guinea Pig genes (cavPor3)                 \n",
       "3  gaculeatus_gene_ensembl        Stickleback genes (BROAD S1)               \n",
       "4  lafricana_gene_ensembl         Elephant genes (Loxafr3.0)                 \n",
       "5  itridecemlineatus_gene_ensembl Squirrel genes (spetri2)                   \n",
       "6  choffmanni_gene_ensembl        Sloth genes (choHof1)                      \n",
       "7  fcatus_gene_ensembl            Cat genes (Felis_catus_6.2)                \n",
       "8  csavignyi_gene_ensembl         C.savignyi genes (CSAV 2.0)                \n",
       "9  rnorvegicus_gene_ensembl       Rat genes (Rnor_6.0)                       \n",
       "10 psinensis_gene_ensembl         Chinese softshell turtle genes (PelSin_1.0)\n",
       "11 cjacchus_gene_ensembl          Marmoset genes (C_jacchus3.2.1)            \n",
       "12 ttruncatus_gene_ensembl        Dolphin genes (turTru1)                    \n",
       "13 scerevisiae_gene_ensembl       Saccharomyces cerevisiae genes (R64-1-1)   \n",
       "14 celegans_gene_ensembl          Caenorhabditis elegans genes (WBcel235)    \n",
       "15 csabaeus_gene_ensembl          Vervet-AGM genes (ChlSab1.1)               \n",
       "16 oniloticus_gene_ensembl        Tilapia genes (Orenil1.0)                  \n",
       "17 trubripes_gene_ensembl         Fugu genes (FUGU 4.0)                      \n",
       "18 amexicanus_gene_ensembl        Cave fish genes (AstMex102)                \n",
       "19 pmarinus_gene_ensembl          Lamprey genes (Pmarinus_7.0)               \n",
       "20 eeuropaeus_gene_ensembl        Hedgehog genes (eriEur1)                   \n",
       "21 falbicollis_gene_ensembl       Flycatcher genes (FicAlb_1.4)              \n",
       "22 ptroglodytes_gene_ensembl      Chimpanzee genes (CHIMP2.1.4)              \n",
       "23 cintestinalis_gene_ensembl     C.intestinalis genes (KH)                  \n",
       "24 etelfairi_gene_ensembl         Lesser hedgehog tenrec genes (TENREC)      \n",
       "25 nleucogenys_gene_ensembl       Gibbon genes (Nleu1.0)                     \n",
       "26 sscrofa_gene_ensembl           Pig genes (Sscrofa10.2)                    \n",
       "27 ocuniculus_gene_ensembl        Rabbit genes (OryCun2.0)                   \n",
       "28 dnovemcinctus_gene_ensembl     Armadillo genes (Dasnov3.0)                \n",
       "29 pcapensis_gene_ensembl         Hyrax genes (proCap1)                      \n",
       "30 tguttata_gene_ensembl          Zebra Finch genes (taeGut3.2.4)            \n",
       "⋮  ⋮                              ⋮                                          \n",
       "40 xmaculatus_gene_ensembl        Platyfish genes (Xipmac4.4.2)              \n",
       "41 drerio_gene_ensembl            Zebrafish genes (GRCz10)                   \n",
       "42 tnigroviridis_gene_ensembl     Tetraodon genes (TETRAODON 8.0)            \n",
       "43 lchalumnae_gene_ensembl        Coelacanth genes (LatCha1)                 \n",
       "44 amelanoleuca_gene_ensembl      Panda genes (ailMel1)                      \n",
       "45 mmulatta_gene_ensembl          Macaque genes (Mmul_8.0.1)                 \n",
       "46 pvampyrus_gene_ensembl         Megabat genes (pteVam1)                    \n",
       "47 panubis_gene_ensembl           Olive baboon genes (PapAnu2.0)             \n",
       "48 mdomestica_gene_ensembl        Opossum genes (monDom5)                    \n",
       "49 acarolinensis_gene_ensembl     Anole lizard genes (AnoCar2.0)             \n",
       "50 vpacos_gene_ensembl            Alpaca genes (vicPac1)                     \n",
       "51 tsyrichta_gene_ensembl         Tarsier genes (tarSyr1)                    \n",
       "52 ogarnettii_gene_ensembl        Bushbaby genes (OtoGar3)                   \n",
       "53 dmelanogaster_gene_ensembl     Fruitfly genes (BDGP6)                     \n",
       "54 loculatus_gene_ensembl         Spotted gar genes (LepOcu1)                \n",
       "55 mmurinus_gene_ensembl          Mouse Lemur genes (Mmur_2.0)               \n",
       "56 olatipes_gene_ensembl          Medaka genes (HdrR)                        \n",
       "57 oprinceps_gene_ensembl         Pika genes (OchPri2.0-Ens)                 \n",
       "58 ggorilla_gene_ensembl          Gorilla genes (gorGor3.1)                  \n",
       "59 dordii_gene_ensembl            Kangaroo rat genes (dipOrd1)               \n",
       "60 oaries_gene_ensembl            Sheep genes (Oar_v3.1)                     \n",
       "61 mmusculus_gene_ensembl         Mouse genes (GRCm38.p5)                    \n",
       "62 mgallopavo_gene_ensembl        Turkey genes (Turkey_2.01)                 \n",
       "63 gmorhua_gene_ensembl           Cod genes (gadMor1)                        \n",
       "64 saraneus_gene_ensembl          Shrew genes (sorAra1)                      \n",
       "65 aplatyrhynchos_gene_ensembl    Duck genes (BGI_duck_1.0)                  \n",
       "66 sharrisii_gene_ensembl         Tasmanian devil genes (Devil_ref v7.0)     \n",
       "67 btaurus_gene_ensembl           Cow genes (UMD3.1)                         \n",
       "68 meugenii_gene_ensembl          Wallaby genes (Meug_1.0)                   \n",
       "69 cfamiliaris_gene_ensembl       Dog genes (CanFam3.1)                      \n",
       "   version        \n",
       "1  OANA5          \n",
       "2  cavPor3        \n",
       "3  BROAD S1       \n",
       "4  Loxafr3.0      \n",
       "5  spetri2        \n",
       "6  choHof1        \n",
       "7  Felis_catus_6.2\n",
       "8  CSAV 2.0       \n",
       "9  Rnor_6.0       \n",
       "10 PelSin_1.0     \n",
       "11 C_jacchus3.2.1 \n",
       "12 turTru1        \n",
       "13 R64-1-1        \n",
       "14 WBcel235       \n",
       "15 ChlSab1.1      \n",
       "16 Orenil1.0      \n",
       "17 FUGU 4.0       \n",
       "18 AstMex102      \n",
       "19 Pmarinus_7.0   \n",
       "20 eriEur1        \n",
       "21 FicAlb_1.4     \n",
       "22 CHIMP2.1.4     \n",
       "23 KH             \n",
       "24 TENREC         \n",
       "25 Nleu1.0        \n",
       "26 Sscrofa10.2    \n",
       "27 OryCun2.0      \n",
       "28 Dasnov3.0      \n",
       "29 proCap1        \n",
       "30 taeGut3.2.4    \n",
       "⋮  ⋮              \n",
       "40 Xipmac4.4.2    \n",
       "41 GRCz10         \n",
       "42 TETRAODON 8.0  \n",
       "43 LatCha1        \n",
       "44 ailMel1        \n",
       "45 Mmul_8.0.1     \n",
       "46 pteVam1        \n",
       "47 PapAnu2.0      \n",
       "48 monDom5        \n",
       "49 AnoCar2.0      \n",
       "50 vicPac1        \n",
       "51 tarSyr1        \n",
       "52 OtoGar3        \n",
       "53 BDGP6          \n",
       "54 LepOcu1        \n",
       "55 Mmur_2.0       \n",
       "56 HdrR           \n",
       "57 OchPri2.0-Ens  \n",
       "58 gorGor3.1      \n",
       "59 dipOrd1        \n",
       "60 Oar_v3.1       \n",
       "61 GRCm38.p5      \n",
       "62 Turkey_2.01    \n",
       "63 gadMor1        \n",
       "64 sorAra1        \n",
       "65 BGI_duck_1.0   \n",
       "66 Devil_ref v7.0 \n",
       "67 UMD3.1         \n",
       "68 Meug_1.0       \n",
       "69 CanFam3.1      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listDatasets(ensembl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ensembl = useDataset(\"hsapiens_gene_ensembl\",mart=ensembl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>name</th><th scope=col>description</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>chromosome_name                          </td><td>Chromosome/scaffold name                 </td></tr>\n",
       "\t<tr><td>start                                    </td><td>Start                                    </td></tr>\n",
       "\t<tr><td>end                                      </td><td>End                                      </td></tr>\n",
       "\t<tr><td>band_start                               </td><td>Band Start                               </td></tr>\n",
       "\t<tr><td>band_end                                 </td><td>Band End                                 </td></tr>\n",
       "\t<tr><td>marker_start                             </td><td>Marker Start                             </td></tr>\n",
       "\t<tr><td>marker_end                               </td><td>Marker End                               </td></tr>\n",
       "\t<tr><td>encode_region                            </td><td>Encode region                            </td></tr>\n",
       "\t<tr><td>strand                                   </td><td>Strand                                   </td></tr>\n",
       "\t<tr><td>chromosomal_region                       </td><td>e.g. 1:100:10000:-1, 1:100000:200000:1   </td></tr>\n",
       "\t<tr><td>with_ccds                                </td><td>With CCDS                                </td></tr>\n",
       "\t<tr><td>with_chembl                              </td><td>With ChEMBL                              </td></tr>\n",
       "\t<tr><td>with_clone_based_ensembl_gene            </td><td>With Clone-based (Ensembl) gene          </td></tr>\n",
       "\t<tr><td>with_clone_based_ensembl_transcript      </td><td>With Clone-based (Ensembl) transcript    </td></tr>\n",
       "\t<tr><td>with_clone_based_vega_gene               </td><td>With Clone-based (Vega) gene             </td></tr>\n",
       "\t<tr><td>with_clone_based_vega_transcript         </td><td>With Clone-based (Vega) transcript       </td></tr>\n",
       "\t<tr><td>with_dbass3                              </td><td>With DataBase of Aberrant 3' Splice Sites</td></tr>\n",
       "\t<tr><td>with_dbass5                              </td><td>With DataBase of Aberrant 5' Splice Sites</td></tr>\n",
       "\t<tr><td>with_ens_hs_gene                         </td><td>With Ensembl Human Gene                  </td></tr>\n",
       "\t<tr><td>with_ens_hs_transcript                   </td><td>With Ensembl Human Transcript            </td></tr>\n",
       "\t<tr><td>with_ens_hs_translation                  </td><td>With Ensembl Human Translation           </td></tr>\n",
       "\t<tr><td>with_entrezgene_trans_name               </td><td>With EntrezGene transcript name          </td></tr>\n",
       "\t<tr><td>with_embl                                </td><td>With European Nucleotide Archive         </td></tr>\n",
       "\t<tr><td>with_arrayexpress                        </td><td>With Expression Atlas                    </td></tr>\n",
       "\t<tr><td>with_genedb                              </td><td>With GeneDB                              </td></tr>\n",
       "\t<tr><td>with_go                                  </td><td>With GO                                  </td></tr>\n",
       "\t<tr><td>with_goslim_goa                          </td><td>With GOSlim GOA                          </td></tr>\n",
       "\t<tr><td>with_ottg                                </td><td>With Havana gene                         </td></tr>\n",
       "\t<tr><td>with_ottt                                </td><td>With Havana transcript                   </td></tr>\n",
       "\t<tr><td>with_ottp                                </td><td>With Havana translation                  </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>with_hmmpanther                                  </td><td>With hmmpanther ID(s)                            </td></tr>\n",
       "\t<tr><td>with_pfam                                        </td><td>With Pfam domain ID(s)                           </td></tr>\n",
       "\t<tr><td>with_pirsf                                       </td><td>With PIRSF domain ID(s)                          </td></tr>\n",
       "\t<tr><td>with_prints                                      </td><td>With Prints domain ID(s)                         </td></tr>\n",
       "\t<tr><td>with_blastprodom                                 </td><td>With ProDom                                      </td></tr>\n",
       "\t<tr><td>with_scanprosite                                 </td><td>With PROSITE patterns ID(s)                      </td></tr>\n",
       "\t<tr><td>with_pfscan                                      </td><td>With PROSITE profiles ID(s)                      </td></tr>\n",
       "\t<tr><td>with_seg                                         </td><td>With Low complexity (Seg)                        </td></tr>\n",
       "\t<tr><td>with_signalp                                     </td><td>With Cleavage site (Signalp)                     </td></tr>\n",
       "\t<tr><td>with_smart                                       </td><td>With SMART domains ID(s)                         </td></tr>\n",
       "\t<tr><td>with_superfamily                                 </td><td>With Superfamily domains ID(s)                   </td></tr>\n",
       "\t<tr><td>with_tigrfam                                     </td><td>With TIGRFAM domain ID(s)                        </td></tr>\n",
       "\t<tr><td>with_tmhmm                                       </td><td>With Transmembrane helices                       </td></tr>\n",
       "\t<tr><td>interpro                                         </td><td>Interpro ID(s) [e.g. IPR000001]                  </td></tr>\n",
       "\t<tr><td>family                                           </td><td>Ensembl Protein Family ID(s) [e.g. PTHR10000_SF7]</td></tr>\n",
       "\t<tr><td>gene3d                                           </td><td>Gene3D ID(s) [e.g. 1.10.10.10]                   </td></tr>\n",
       "\t<tr><td>hamap                                            </td><td>HAMAP ID(s) [e.g. MF_00001]                      </td></tr>\n",
       "\t<tr><td>hmmpanther                                       </td><td>hmmpanther ID(s) [e.g. PTHR10000]                </td></tr>\n",
       "\t<tr><td>pfam                                             </td><td>Pfam domain ID(s) [e.g. PF00001]                 </td></tr>\n",
       "\t<tr><td>pirsf                                            </td><td>PIRSF domain ID(s) [e.g. PIRSF000019]            </td></tr>\n",
       "\t<tr><td>prints                                           </td><td>Prints domain ID(s) [e.g. PR00001]               </td></tr>\n",
       "\t<tr><td>scanprosite                                      </td><td>PROSITE patterns ID(s) [e.g. PS00010]            </td></tr>\n",
       "\t<tr><td>pfscan                                           </td><td>PROSITE profiles ID(s) [e.g. PS01031]            </td></tr>\n",
       "\t<tr><td>smart                                            </td><td>SMART domains ID(s) [e.g. SM00002]               </td></tr>\n",
       "\t<tr><td>superfamily                                      </td><td>Superfamily domains ID(s) [e.g. 0034913]         </td></tr>\n",
       "\t<tr><td>tigrfam                                          </td><td>TIGRFAM domain ID(s) [e.g. TIGR00002]            </td></tr>\n",
       "\t<tr><td>germ_line_variation_source                       </td><td>limit to genes with germline variant data sources</td></tr>\n",
       "\t<tr><td>somatic_variation_source                         </td><td>limit to genes with somatic variant sources      </td></tr>\n",
       "\t<tr><td>with_validated_snp                               </td><td>Variant supporting evidence                      </td></tr>\n",
       "\t<tr><td>so_mini_parent_name                              </td><td>Parent term name                                 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       " name & description\\\\\n",
       "\\hline\n",
       "\t chromosome\\_name                         & Chromosome/scaffold name                 \\\\\n",
       "\t start                                     & Start                                    \\\\\n",
       "\t end                                       & End                                      \\\\\n",
       "\t band\\_start                              & Band Start                               \\\\\n",
       "\t band\\_end                                & Band End                                 \\\\\n",
       "\t marker\\_start                            & Marker Start                             \\\\\n",
       "\t marker\\_end                              & Marker End                               \\\\\n",
       "\t encode\\_region                           & Encode region                            \\\\\n",
       "\t strand                                    & Strand                                   \\\\\n",
       "\t chromosomal\\_region                      & e.g. 1:100:10000:-1, 1:100000:200000:1   \\\\\n",
       "\t with\\_ccds                               & With CCDS                                \\\\\n",
       "\t with\\_chembl                             & With ChEMBL                              \\\\\n",
       "\t with\\_clone\\_based\\_ensembl\\_gene       & With Clone-based (Ensembl) gene            \\\\\n",
       "\t with\\_clone\\_based\\_ensembl\\_transcript & With Clone-based (Ensembl) transcript      \\\\\n",
       "\t with\\_clone\\_based\\_vega\\_gene          & With Clone-based (Vega) gene               \\\\\n",
       "\t with\\_clone\\_based\\_vega\\_transcript    & With Clone-based (Vega) transcript         \\\\\n",
       "\t with\\_dbass3                             & With DataBase of Aberrant 3' Splice Sites\\\\\n",
       "\t with\\_dbass5                             & With DataBase of Aberrant 5' Splice Sites\\\\\n",
       "\t with\\_ens\\_hs\\_gene                    & With Ensembl Human Gene                  \\\\\n",
       "\t with\\_ens\\_hs\\_transcript              & With Ensembl Human Transcript            \\\\\n",
       "\t with\\_ens\\_hs\\_translation             & With Ensembl Human Translation           \\\\\n",
       "\t with\\_entrezgene\\_trans\\_name          & With EntrezGene transcript name          \\\\\n",
       "\t with\\_embl                               & With European Nucleotide Archive         \\\\\n",
       "\t with\\_arrayexpress                       & With Expression Atlas                    \\\\\n",
       "\t with\\_genedb                             & With GeneDB                              \\\\\n",
       "\t with\\_go                                 & With GO                                  \\\\\n",
       "\t with\\_goslim\\_goa                       & With GOSlim GOA                          \\\\\n",
       "\t with\\_ottg                               & With Havana gene                         \\\\\n",
       "\t with\\_ottt                               & With Havana transcript                   \\\\\n",
       "\t with\\_ottp                               & With Havana translation                  \\\\\n",
       "\t ⋮ & ⋮\\\\\n",
       "\t with\\_hmmpanther                                 & With hmmpanther ID(s)                            \\\\\n",
       "\t with\\_pfam                                       & With Pfam domain ID(s)                           \\\\\n",
       "\t with\\_pirsf                                      & With PIRSF domain ID(s)                          \\\\\n",
       "\t with\\_prints                                     & With Prints domain ID(s)                         \\\\\n",
       "\t with\\_blastprodom                                & With ProDom                                      \\\\\n",
       "\t with\\_scanprosite                                & With PROSITE patterns ID(s)                      \\\\\n",
       "\t with\\_pfscan                                     & With PROSITE profiles ID(s)                      \\\\\n",
       "\t with\\_seg                                        & With Low complexity (Seg)                        \\\\\n",
       "\t with\\_signalp                                    & With Cleavage site (Signalp)                     \\\\\n",
       "\t with\\_smart                                      & With SMART domains ID(s)                         \\\\\n",
       "\t with\\_superfamily                                & With Superfamily domains ID(s)                   \\\\\n",
       "\t with\\_tigrfam                                    & With TIGRFAM domain ID(s)                        \\\\\n",
       "\t with\\_tmhmm                                      & With Transmembrane helices                       \\\\\n",
       "\t interpro                                              & Interpro ID(s) {[}e.g. IPR000001{]}                  \\\\\n",
       "\t family                                                  & Ensembl Protein Family ID(s) {[}e.g. PTHR10000\\_SF7{]}\\\\\n",
       "\t gene3d                                                & Gene3D ID(s) {[}e.g. 1.10.10.10{]}                   \\\\\n",
       "\t hamap                                                   & HAMAP ID(s) {[}e.g. MF\\_00001{]}                      \\\\\n",
       "\t hmmpanther                                            & hmmpanther ID(s) {[}e.g. PTHR10000{]}                \\\\\n",
       "\t pfam                                                  & Pfam domain ID(s) {[}e.g. PF00001{]}                 \\\\\n",
       "\t pirsf                                                 & PIRSF domain ID(s) {[}e.g. PIRSF000019{]}            \\\\\n",
       "\t prints                                                & Prints domain ID(s) {[}e.g. PR00001{]}               \\\\\n",
       "\t scanprosite                                           & PROSITE patterns ID(s) {[}e.g. PS00010{]}            \\\\\n",
       "\t pfscan                                                & PROSITE profiles ID(s) {[}e.g. PS01031{]}            \\\\\n",
       "\t smart                                                 & SMART domains ID(s) {[}e.g. SM00002{]}               \\\\\n",
       "\t superfamily                                           & Superfamily domains ID(s) {[}e.g. 0034913{]}         \\\\\n",
       "\t tigrfam                                               & TIGRFAM domain ID(s) {[}e.g. TIGR00002{]}            \\\\\n",
       "\t germ\\_line\\_variation\\_source                  & limit to genes with germline variant data sources\\\\\n",
       "\t somatic\\_variation\\_source                      & limit to genes with somatic variant sources      \\\\\n",
       "\t with\\_validated\\_snp                            & Variant supporting evidence                      \\\\\n",
       "\t so\\_mini\\_parent\\_name                         & Parent term name                                 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "name | description | \n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| chromosome_name                           | Chromosome/scaffold name                  | \n",
       "| start                                     | Start                                     | \n",
       "| end                                       | End                                       | \n",
       "| band_start                                | Band Start                                | \n",
       "| band_end                                  | Band End                                  | \n",
       "| marker_start                              | Marker Start                              | \n",
       "| marker_end                                | Marker End                                | \n",
       "| encode_region                             | Encode region                             | \n",
       "| strand                                    | Strand                                    | \n",
       "| chromosomal_region                        | e.g. 1:100:10000:-1, 1:100000:200000:1    | \n",
       "| with_ccds                                 | With CCDS                                 | \n",
       "| with_chembl                               | With ChEMBL                               | \n",
       "| with_clone_based_ensembl_gene             | With Clone-based (Ensembl) gene           | \n",
       "| with_clone_based_ensembl_transcript       | With Clone-based (Ensembl) transcript     | \n",
       "| with_clone_based_vega_gene                | With Clone-based (Vega) gene              | \n",
       "| with_clone_based_vega_transcript          | With Clone-based (Vega) transcript        | \n",
       "| with_dbass3                               | With DataBase of Aberrant 3' Splice Sites | \n",
       "| with_dbass5                               | With DataBase of Aberrant 5' Splice Sites | \n",
       "| with_ens_hs_gene                          | With Ensembl Human Gene                   | \n",
       "| with_ens_hs_transcript                    | With Ensembl Human Transcript             | \n",
       "| with_ens_hs_translation                   | With Ensembl Human Translation            | \n",
       "| with_entrezgene_trans_name                | With EntrezGene transcript name           | \n",
       "| with_embl                                 | With European Nucleotide Archive          | \n",
       "| with_arrayexpress                         | With Expression Atlas                     | \n",
       "| with_genedb                               | With GeneDB                               | \n",
       "| with_go                                   | With GO                                   | \n",
       "| with_goslim_goa                           | With GOSlim GOA                           | \n",
       "| with_ottg                                 | With Havana gene                          | \n",
       "| with_ottt                                 | With Havana transcript                    | \n",
       "| with_ottp                                 | With Havana translation                   | \n",
       "| ⋮ | ⋮ | \n",
       "| with_hmmpanther                                   | With hmmpanther ID(s)                             | \n",
       "| with_pfam                                         | With Pfam domain ID(s)                            | \n",
       "| with_pirsf                                        | With PIRSF domain ID(s)                           | \n",
       "| with_prints                                       | With Prints domain ID(s)                          | \n",
       "| with_blastprodom                                  | With ProDom                                       | \n",
       "| with_scanprosite                                  | With PROSITE patterns ID(s)                       | \n",
       "| with_pfscan                                       | With PROSITE profiles ID(s)                       | \n",
       "| with_seg                                          | With Low complexity (Seg)                         | \n",
       "| with_signalp                                      | With Cleavage site (Signalp)                      | \n",
       "| with_smart                                        | With SMART domains ID(s)                          | \n",
       "| with_superfamily                                  | With Superfamily domains ID(s)                    | \n",
       "| with_tigrfam                                      | With TIGRFAM domain ID(s)                         | \n",
       "| with_tmhmm                                        | With Transmembrane helices                        | \n",
       "| interpro                                          | Interpro ID(s) [e.g. IPR000001]                   | \n",
       "| family                                            | Ensembl Protein Family ID(s) [e.g. PTHR10000_SF7] | \n",
       "| gene3d                                            | Gene3D ID(s) [e.g. 1.10.10.10]                    | \n",
       "| hamap                                             | HAMAP ID(s) [e.g. MF_00001]                       | \n",
       "| hmmpanther                                        | hmmpanther ID(s) [e.g. PTHR10000]                 | \n",
       "| pfam                                              | Pfam domain ID(s) [e.g. PF00001]                  | \n",
       "| pirsf                                             | PIRSF domain ID(s) [e.g. PIRSF000019]             | \n",
       "| prints                                            | Prints domain ID(s) [e.g. PR00001]                | \n",
       "| scanprosite                                       | PROSITE patterns ID(s) [e.g. PS00010]             | \n",
       "| pfscan                                            | PROSITE profiles ID(s) [e.g. PS01031]             | \n",
       "| smart                                             | SMART domains ID(s) [e.g. SM00002]                | \n",
       "| superfamily                                       | Superfamily domains ID(s) [e.g. 0034913]          | \n",
       "| tigrfam                                           | TIGRFAM domain ID(s) [e.g. TIGR00002]             | \n",
       "| germ_line_variation_source                        | limit to genes with germline variant data sources | \n",
       "| somatic_variation_source                          | limit to genes with somatic variant sources       | \n",
       "| with_validated_snp                                | Variant supporting evidence                       | \n",
       "| so_mini_parent_name                               | Parent term name                                  | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "    name                               \n",
       "1   chromosome_name                    \n",
       "2   start                              \n",
       "3   end                                \n",
       "4   band_start                         \n",
       "5   band_end                           \n",
       "6   marker_start                       \n",
       "7   marker_end                         \n",
       "8   encode_region                      \n",
       "9   strand                             \n",
       "10  chromosomal_region                 \n",
       "11  with_ccds                          \n",
       "12  with_chembl                        \n",
       "13  with_clone_based_ensembl_gene      \n",
       "14  with_clone_based_ensembl_transcript\n",
       "15  with_clone_based_vega_gene         \n",
       "16  with_clone_based_vega_transcript   \n",
       "17  with_dbass3                        \n",
       "18  with_dbass5                        \n",
       "19  with_ens_hs_gene                   \n",
       "20  with_ens_hs_transcript             \n",
       "21  with_ens_hs_translation            \n",
       "22  with_entrezgene_trans_name         \n",
       "23  with_embl                          \n",
       "24  with_arrayexpress                  \n",
       "25  with_genedb                        \n",
       "26  with_go                            \n",
       "27  with_goslim_goa                    \n",
       "28  with_ottg                          \n",
       "29  with_ottt                          \n",
       "30  with_ottp                          \n",
       "⋮   ⋮                                  \n",
       "296 with_hmmpanther                    \n",
       "297 with_pfam                          \n",
       "298 with_pirsf                         \n",
       "299 with_prints                        \n",
       "300 with_blastprodom                   \n",
       "301 with_scanprosite                   \n",
       "302 with_pfscan                        \n",
       "303 with_seg                           \n",
       "304 with_signalp                       \n",
       "305 with_smart                         \n",
       "306 with_superfamily                   \n",
       "307 with_tigrfam                       \n",
       "308 with_tmhmm                         \n",
       "309 interpro                           \n",
       "310 family                             \n",
       "311 gene3d                             \n",
       "312 hamap                              \n",
       "313 hmmpanther                         \n",
       "314 pfam                               \n",
       "315 pirsf                              \n",
       "316 prints                             \n",
       "317 scanprosite                        \n",
       "318 pfscan                             \n",
       "319 smart                              \n",
       "320 superfamily                        \n",
       "321 tigrfam                            \n",
       "322 germ_line_variation_source         \n",
       "323 somatic_variation_source           \n",
       "324 with_validated_snp                 \n",
       "325 so_mini_parent_name                \n",
       "    description                                      \n",
       "1   Chromosome/scaffold name                         \n",
       "2   Start                                            \n",
       "3   End                                              \n",
       "4   Band Start                                       \n",
       "5   Band End                                         \n",
       "6   Marker Start                                     \n",
       "7   Marker End                                       \n",
       "8   Encode region                                    \n",
       "9   Strand                                           \n",
       "10  e.g. 1:100:10000:-1, 1:100000:200000:1           \n",
       "11  With CCDS                                        \n",
       "12  With ChEMBL                                      \n",
       "13  With Clone-based (Ensembl) gene                  \n",
       "14  With Clone-based (Ensembl) transcript            \n",
       "15  With Clone-based (Vega) gene                     \n",
       "16  With Clone-based (Vega) transcript               \n",
       "17  With DataBase of Aberrant 3' Splice Sites        \n",
       "18  With DataBase of Aberrant 5' Splice Sites        \n",
       "19  With Ensembl Human Gene                          \n",
       "20  With Ensembl Human Transcript                    \n",
       "21  With Ensembl Human Translation                   \n",
       "22  With EntrezGene transcript name                  \n",
       "23  With European Nucleotide Archive                 \n",
       "24  With Expression Atlas                            \n",
       "25  With GeneDB                                      \n",
       "26  With GO                                          \n",
       "27  With GOSlim GOA                                  \n",
       "28  With Havana gene                                 \n",
       "29  With Havana transcript                           \n",
       "30  With Havana translation                          \n",
       "⋮   ⋮                                                \n",
       "296 With hmmpanther ID(s)                            \n",
       "297 With Pfam domain ID(s)                           \n",
       "298 With PIRSF domain ID(s)                          \n",
       "299 With Prints domain ID(s)                         \n",
       "300 With ProDom                                      \n",
       "301 With PROSITE patterns ID(s)                      \n",
       "302 With PROSITE profiles ID(s)                      \n",
       "303 With Low complexity (Seg)                        \n",
       "304 With Cleavage site (Signalp)                     \n",
       "305 With SMART domains ID(s)                         \n",
       "306 With Superfamily domains ID(s)                   \n",
       "307 With TIGRFAM domain ID(s)                        \n",
       "308 With Transmembrane helices                       \n",
       "309 Interpro ID(s) [e.g. IPR000001]                  \n",
       "310 Ensembl Protein Family ID(s) [e.g. PTHR10000_SF7]\n",
       "311 Gene3D ID(s) [e.g. 1.10.10.10]                   \n",
       "312 HAMAP ID(s) [e.g. MF_00001]                      \n",
       "313 hmmpanther ID(s) [e.g. PTHR10000]                \n",
       "314 Pfam domain ID(s) [e.g. PF00001]                 \n",
       "315 PIRSF domain ID(s) [e.g. PIRSF000019]            \n",
       "316 Prints domain ID(s) [e.g. PR00001]               \n",
       "317 PROSITE patterns ID(s) [e.g. PS00010]            \n",
       "318 PROSITE profiles ID(s) [e.g. PS01031]            \n",
       "319 SMART domains ID(s) [e.g. SM00002]               \n",
       "320 Superfamily domains ID(s) [e.g. 0034913]         \n",
       "321 TIGRFAM domain ID(s) [e.g. TIGR00002]            \n",
       "322 limit to genes with germline variant data sources\n",
       "323 limit to genes with somatic variant sources      \n",
       "324 Variant supporting evidence                      \n",
       "325 Parent term name                                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filters = listFilters(ensembl)\n",
    "filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>name</th><th scope=col>description</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>ensembl_gene_id                           </td><td>Gene stable ID                            </td></tr>\n",
       "\t<tr><td>ensembl_transcript_id                     </td><td>Transcript stable ID                      </td></tr>\n",
       "\t<tr><td>ensembl_peptide_id                        </td><td>Protein stable ID                         </td></tr>\n",
       "\t<tr><td>ensembl_exon_id                           </td><td>Exon stable ID                            </td></tr>\n",
       "\t<tr><td>description                               </td><td>Gene description                          </td></tr>\n",
       "\t<tr><td>chromosome_name                           </td><td>Chromosome/scaffold name                  </td></tr>\n",
       "\t<tr><td>start_position                            </td><td>Gene Start (bp)                           </td></tr>\n",
       "\t<tr><td>end_position                              </td><td>Gene End (bp)                             </td></tr>\n",
       "\t<tr><td>strand                                    </td><td>Strand                                    </td></tr>\n",
       "\t<tr><td>band                                      </td><td>Karyotype band                            </td></tr>\n",
       "\t<tr><td>transcript_start                          </td><td>Transcript start (bp)                     </td></tr>\n",
       "\t<tr><td>transcript_end                            </td><td>Transcript end (bp)                       </td></tr>\n",
       "\t<tr><td>transcription_start_site                  </td><td>Transcription start site (TSS)            </td></tr>\n",
       "\t<tr><td>transcript_length                         </td><td>Transcript length (including UTRs and CDS)</td></tr>\n",
       "\t<tr><td>transcript_tsl                            </td><td>Transcript support level (TSL)            </td></tr>\n",
       "\t<tr><td>transcript_gencode_basic                  </td><td>GENCODE basic annotation                  </td></tr>\n",
       "\t<tr><td>transcript_appris                         </td><td>APPRIS annotation                         </td></tr>\n",
       "\t<tr><td>external_gene_name                        </td><td>Gene name                                 </td></tr>\n",
       "\t<tr><td>external_gene_source                      </td><td>Source of gene name                       </td></tr>\n",
       "\t<tr><td>external_transcript_name                  </td><td>Transcript name                           </td></tr>\n",
       "\t<tr><td>external_transcript_source_name           </td><td>Source of transcript name                 </td></tr>\n",
       "\t<tr><td>transcript_count                          </td><td>Transcript count                          </td></tr>\n",
       "\t<tr><td>percentage_gene_gc_content                </td><td>% GC content                              </td></tr>\n",
       "\t<tr><td>gene_biotype                              </td><td>Gene type                                 </td></tr>\n",
       "\t<tr><td>transcript_biotype                        </td><td>Transcript type                           </td></tr>\n",
       "\t<tr><td>source                                    </td><td>Source (gene)                             </td></tr>\n",
       "\t<tr><td>transcript_source                         </td><td>Source (transcript)                       </td></tr>\n",
       "\t<tr><td>status                                    </td><td>Status (gene)                             </td></tr>\n",
       "\t<tr><td>transcript_status                         </td><td>Status (transcript)                       </td></tr>\n",
       "\t<tr><td>version                                   </td><td>Version (gene)                            </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>uniprotsptrembl                           </td><td>UniProtKB/TrEMBL ID                       </td></tr>\n",
       "\t<tr><td>cdna_coding_start                         </td><td>CDS start (within cDNA)                   </td></tr>\n",
       "\t<tr><td>cdna_coding_end                           </td><td>CDS end (within cDNA)                     </td></tr>\n",
       "\t<tr><td>5_utr_start                               </td><td>5' UTR start                              </td></tr>\n",
       "\t<tr><td>5_utr_end                                 </td><td>5' UTR end                                </td></tr>\n",
       "\t<tr><td>3_utr_start                               </td><td>3' UTR start                              </td></tr>\n",
       "\t<tr><td>3_utr_end                                 </td><td>3' UTR end                                </td></tr>\n",
       "\t<tr><td>ensembl_transcript_id                     </td><td>Transcript stable ID                      </td></tr>\n",
       "\t<tr><td>ensembl_peptide_id                        </td><td>Protein stable ID                         </td></tr>\n",
       "\t<tr><td>transcript_biotype                        </td><td>Transcript type                           </td></tr>\n",
       "\t<tr><td>strand                                    </td><td>Strand                                    </td></tr>\n",
       "\t<tr><td>transcript_start                          </td><td>Transcript start (bp)                     </td></tr>\n",
       "\t<tr><td>transcript_end                            </td><td>Transcript end (bp)                       </td></tr>\n",
       "\t<tr><td>transcription_start_site                  </td><td>Transcription start site (TSS)            </td></tr>\n",
       "\t<tr><td>transcript_length                         </td><td>Transcript length (including UTRs and CDS)</td></tr>\n",
       "\t<tr><td>cds_length                                </td><td>CDS Length                                </td></tr>\n",
       "\t<tr><td>cds_start                                 </td><td>CDS start                                 </td></tr>\n",
       "\t<tr><td>cds_end                                   </td><td>CDS end                                   </td></tr>\n",
       "\t<tr><td>ensembl_exon_id                           </td><td>Exon stable ID                            </td></tr>\n",
       "\t<tr><td>exon_chrom_start                          </td><td>Exon region start (bp)                    </td></tr>\n",
       "\t<tr><td>exon_chrom_end                            </td><td>Exon region end (bp)                      </td></tr>\n",
       "\t<tr><td>strand                                    </td><td>Strand                                    </td></tr>\n",
       "\t<tr><td>rank                                      </td><td>Exon Rank in Transcript                   </td></tr>\n",
       "\t<tr><td>phase                                     </td><td>start phase                               </td></tr>\n",
       "\t<tr><td>end_phase                                 </td><td>end phase                                 </td></tr>\n",
       "\t<tr><td>cdna_coding_start                         </td><td>cDNA coding start                         </td></tr>\n",
       "\t<tr><td>cdna_coding_end                           </td><td>cDNA coding end                           </td></tr>\n",
       "\t<tr><td>genomic_coding_start                      </td><td>Genomic coding start                      </td></tr>\n",
       "\t<tr><td>genomic_coding_end                        </td><td>Genomic coding end                        </td></tr>\n",
       "\t<tr><td>is_constitutive                           </td><td>Constitutive Exon                         </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       " name & description\\\\\n",
       "\\hline\n",
       "\t ensembl\\_gene\\_id                        & Gene stable ID                            \\\\\n",
       "\t ensembl\\_transcript\\_id                  & Transcript stable ID                      \\\\\n",
       "\t ensembl\\_peptide\\_id                     & Protein stable ID                         \\\\\n",
       "\t ensembl\\_exon\\_id                        & Exon stable ID                            \\\\\n",
       "\t description                                & Gene description                          \\\\\n",
       "\t chromosome\\_name                          & Chromosome/scaffold name                  \\\\\n",
       "\t start\\_position                           & Gene Start (bp)                           \\\\\n",
       "\t end\\_position                             & Gene End (bp)                             \\\\\n",
       "\t strand                                     & Strand                                    \\\\\n",
       "\t band                                       & Karyotype band                            \\\\\n",
       "\t transcript\\_start                         & Transcript start (bp)                     \\\\\n",
       "\t transcript\\_end                           & Transcript end (bp)                       \\\\\n",
       "\t transcription\\_start\\_site               & Transcription start site (TSS)            \\\\\n",
       "\t transcript\\_length                        & Transcript length (including UTRs and CDS)\\\\\n",
       "\t transcript\\_tsl                           & Transcript support level (TSL)            \\\\\n",
       "\t transcript\\_gencode\\_basic               & GENCODE basic annotation                  \\\\\n",
       "\t transcript\\_appris                        & APPRIS annotation                         \\\\\n",
       "\t external\\_gene\\_name                     & Gene name                                 \\\\\n",
       "\t external\\_gene\\_source                   & Source of gene name                       \\\\\n",
       "\t external\\_transcript\\_name               & Transcript name                           \\\\\n",
       "\t external\\_transcript\\_source\\_name      & Source of transcript name                 \\\\\n",
       "\t transcript\\_count                         & Transcript count                          \\\\\n",
       "\t percentage\\_gene\\_gc\\_content             & \\% GC content                              \\\\\n",
       "\t gene\\_biotype                             & Gene type                                 \\\\\n",
       "\t transcript\\_biotype                       & Transcript type                           \\\\\n",
       "\t source                                     & Source (gene)                             \\\\\n",
       "\t transcript\\_source                        & Source (transcript)                       \\\\\n",
       "\t status                                     & Status (gene)                             \\\\\n",
       "\t transcript\\_status                        & Status (transcript)                       \\\\\n",
       "\t version                                    & Version (gene)                            \\\\\n",
       "\t ⋮ & ⋮\\\\\n",
       "\t uniprotsptrembl                            & UniProtKB/TrEMBL ID                       \\\\\n",
       "\t cdna\\_coding\\_start                      & CDS start (within cDNA)                   \\\\\n",
       "\t cdna\\_coding\\_end                        & CDS end (within cDNA)                     \\\\\n",
       "\t 5\\_utr\\_start                            & 5' UTR start                              \\\\\n",
       "\t 5\\_utr\\_end                              & 5' UTR end                                \\\\\n",
       "\t 3\\_utr\\_start                            & 3' UTR start                              \\\\\n",
       "\t 3\\_utr\\_end                              & 3' UTR end                                \\\\\n",
       "\t ensembl\\_transcript\\_id                  & Transcript stable ID                      \\\\\n",
       "\t ensembl\\_peptide\\_id                     & Protein stable ID                         \\\\\n",
       "\t transcript\\_biotype                       & Transcript type                           \\\\\n",
       "\t strand                                     & Strand                                    \\\\\n",
       "\t transcript\\_start                         & Transcript start (bp)                     \\\\\n",
       "\t transcript\\_end                           & Transcript end (bp)                       \\\\\n",
       "\t transcription\\_start\\_site               & Transcription start site (TSS)            \\\\\n",
       "\t transcript\\_length                        & Transcript length (including UTRs and CDS)\\\\\n",
       "\t cds\\_length                               & CDS Length                                \\\\\n",
       "\t cds\\_start                                & CDS start                                 \\\\\n",
       "\t cds\\_end                                  & CDS end                                   \\\\\n",
       "\t ensembl\\_exon\\_id                        & Exon stable ID                            \\\\\n",
       "\t exon\\_chrom\\_start                       & Exon region start (bp)                    \\\\\n",
       "\t exon\\_chrom\\_end                         & Exon region end (bp)                      \\\\\n",
       "\t strand                                     & Strand                                    \\\\\n",
       "\t rank                                       & Exon Rank in Transcript                   \\\\\n",
       "\t phase                                      & start phase                               \\\\\n",
       "\t end\\_phase                                & end phase                                 \\\\\n",
       "\t cdna\\_coding\\_start                      & cDNA coding start                         \\\\\n",
       "\t cdna\\_coding\\_end                        & cDNA coding end                           \\\\\n",
       "\t genomic\\_coding\\_start                   & Genomic coding start                      \\\\\n",
       "\t genomic\\_coding\\_end                     & Genomic coding end                        \\\\\n",
       "\t is\\_constitutive                          & Constitutive Exon                         \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "name | description | \n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ensembl_gene_id                            | Gene stable ID                             | \n",
       "| ensembl_transcript_id                      | Transcript stable ID                       | \n",
       "| ensembl_peptide_id                         | Protein stable ID                          | \n",
       "| ensembl_exon_id                            | Exon stable ID                             | \n",
       "| description                                | Gene description                           | \n",
       "| chromosome_name                            | Chromosome/scaffold name                   | \n",
       "| start_position                             | Gene Start (bp)                            | \n",
       "| end_position                               | Gene End (bp)                              | \n",
       "| strand                                     | Strand                                     | \n",
       "| band                                       | Karyotype band                             | \n",
       "| transcript_start                           | Transcript start (bp)                      | \n",
       "| transcript_end                             | Transcript end (bp)                        | \n",
       "| transcription_start_site                   | Transcription start site (TSS)             | \n",
       "| transcript_length                          | Transcript length (including UTRs and CDS) | \n",
       "| transcript_tsl                             | Transcript support level (TSL)             | \n",
       "| transcript_gencode_basic                   | GENCODE basic annotation                   | \n",
       "| transcript_appris                          | APPRIS annotation                          | \n",
       "| external_gene_name                         | Gene name                                  | \n",
       "| external_gene_source                       | Source of gene name                        | \n",
       "| external_transcript_name                   | Transcript name                            | \n",
       "| external_transcript_source_name            | Source of transcript name                  | \n",
       "| transcript_count                           | Transcript count                           | \n",
       "| percentage_gene_gc_content                 | % GC content                               | \n",
       "| gene_biotype                               | Gene type                                  | \n",
       "| transcript_biotype                         | Transcript type                            | \n",
       "| source                                     | Source (gene)                              | \n",
       "| transcript_source                          | Source (transcript)                        | \n",
       "| status                                     | Status (gene)                              | \n",
       "| transcript_status                          | Status (transcript)                        | \n",
       "| version                                    | Version (gene)                             | \n",
       "| ⋮ | ⋮ | \n",
       "| uniprotsptrembl                            | UniProtKB/TrEMBL ID                        | \n",
       "| cdna_coding_start                          | CDS start (within cDNA)                    | \n",
       "| cdna_coding_end                            | CDS end (within cDNA)                      | \n",
       "| 5_utr_start                                | 5' UTR start                               | \n",
       "| 5_utr_end                                  | 5' UTR end                                 | \n",
       "| 3_utr_start                                | 3' UTR start                               | \n",
       "| 3_utr_end                                  | 3' UTR end                                 | \n",
       "| ensembl_transcript_id                      | Transcript stable ID                       | \n",
       "| ensembl_peptide_id                         | Protein stable ID                          | \n",
       "| transcript_biotype                         | Transcript type                            | \n",
       "| strand                                     | Strand                                     | \n",
       "| transcript_start                           | Transcript start (bp)                      | \n",
       "| transcript_end                             | Transcript end (bp)                        | \n",
       "| transcription_start_site                   | Transcription start site (TSS)             | \n",
       "| transcript_length                          | Transcript length (including UTRs and CDS) | \n",
       "| cds_length                                 | CDS Length                                 | \n",
       "| cds_start                                  | CDS start                                  | \n",
       "| cds_end                                    | CDS end                                    | \n",
       "| ensembl_exon_id                            | Exon stable ID                             | \n",
       "| exon_chrom_start                           | Exon region start (bp)                     | \n",
       "| exon_chrom_end                             | Exon region end (bp)                       | \n",
       "| strand                                     | Strand                                     | \n",
       "| rank                                       | Exon Rank in Transcript                    | \n",
       "| phase                                      | start phase                                | \n",
       "| end_phase                                  | end phase                                  | \n",
       "| cdna_coding_start                          | cDNA coding start                          | \n",
       "| cdna_coding_end                            | cDNA coding end                            | \n",
       "| genomic_coding_start                       | Genomic coding start                       | \n",
       "| genomic_coding_end                         | Genomic coding end                         | \n",
       "| is_constitutive                            | Constitutive Exon                          | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "     name                            description                               \n",
       "1    ensembl_gene_id                 Gene stable ID                            \n",
       "2    ensembl_transcript_id           Transcript stable ID                      \n",
       "3    ensembl_peptide_id              Protein stable ID                         \n",
       "4    ensembl_exon_id                 Exon stable ID                            \n",
       "5    description                     Gene description                          \n",
       "6    chromosome_name                 Chromosome/scaffold name                  \n",
       "7    start_position                  Gene Start (bp)                           \n",
       "8    end_position                    Gene End (bp)                             \n",
       "9    strand                          Strand                                    \n",
       "10   band                            Karyotype band                            \n",
       "11   transcript_start                Transcript start (bp)                     \n",
       "12   transcript_end                  Transcript end (bp)                       \n",
       "13   transcription_start_site        Transcription start site (TSS)            \n",
       "14   transcript_length               Transcript length (including UTRs and CDS)\n",
       "15   transcript_tsl                  Transcript support level (TSL)            \n",
       "16   transcript_gencode_basic        GENCODE basic annotation                  \n",
       "17   transcript_appris               APPRIS annotation                         \n",
       "18   external_gene_name              Gene name                                 \n",
       "19   external_gene_source            Source of gene name                       \n",
       "20   external_transcript_name        Transcript name                           \n",
       "21   external_transcript_source_name Source of transcript name                 \n",
       "22   transcript_count                Transcript count                          \n",
       "23   percentage_gene_gc_content      % GC content                              \n",
       "24   gene_biotype                    Gene type                                 \n",
       "25   transcript_biotype              Transcript type                           \n",
       "26   source                          Source (gene)                             \n",
       "27   transcript_source               Source (transcript)                       \n",
       "28   status                          Status (gene)                             \n",
       "29   transcript_status               Status (transcript)                       \n",
       "30   version                         Version (gene)                            \n",
       "⋮    ⋮                               ⋮                                         \n",
       "1448 uniprotsptrembl                 UniProtKB/TrEMBL ID                       \n",
       "1449 cdna_coding_start               CDS start (within cDNA)                   \n",
       "1450 cdna_coding_end                 CDS end (within cDNA)                     \n",
       "1451 5_utr_start                     5' UTR start                              \n",
       "1452 5_utr_end                       5' UTR end                                \n",
       "1453 3_utr_start                     3' UTR start                              \n",
       "1454 3_utr_end                       3' UTR end                                \n",
       "1455 ensembl_transcript_id           Transcript stable ID                      \n",
       "1456 ensembl_peptide_id              Protein stable ID                         \n",
       "1457 transcript_biotype              Transcript type                           \n",
       "1458 strand                          Strand                                    \n",
       "1459 transcript_start                Transcript start (bp)                     \n",
       "1460 transcript_end                  Transcript end (bp)                       \n",
       "1461 transcription_start_site        Transcription start site (TSS)            \n",
       "1462 transcript_length               Transcript length (including UTRs and CDS)\n",
       "1463 cds_length                      CDS Length                                \n",
       "1464 cds_start                       CDS start                                 \n",
       "1465 cds_end                         CDS end                                   \n",
       "1466 ensembl_exon_id                 Exon stable ID                            \n",
       "1467 exon_chrom_start                Exon region start (bp)                    \n",
       "1468 exon_chrom_end                  Exon region end (bp)                      \n",
       "1469 strand                          Strand                                    \n",
       "1470 rank                            Exon Rank in Transcript                   \n",
       "1471 phase                           start phase                               \n",
       "1472 end_phase                       end phase                                 \n",
       "1473 cdna_coding_start               cDNA coding start                         \n",
       "1474 cdna_coding_end                 cDNA coding end                           \n",
       "1475 genomic_coding_start            Genomic coding start                      \n",
       "1476 genomic_coding_end              Genomic coding end                        \n",
       "1477 is_constitutive                 Constitutive Exon                         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "attributes = listAttributes(ensembl)\n",
    "attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>affy_hg_u133_plus_2</th><th scope=col>entrezgene</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>202763_at  </td><td>836        </td></tr>\n",
       "\t<tr><td>209310_s_at</td><td>837        </td></tr>\n",
       "\t<tr><td>207500_at  </td><td>838        </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       " affy\\_hg\\_u133\\_plus\\_2 & entrezgene\\\\\n",
       "\\hline\n",
       "\t 202763\\_at   & 836          \\\\\n",
       "\t 209310\\_s\\_at & 837            \\\\\n",
       "\t 207500\\_at   & 838          \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "affy_hg_u133_plus_2 | entrezgene | \n",
       "|---|---|---|\n",
       "| 202763_at   | 836         | \n",
       "| 209310_s_at | 837         | \n",
       "| 207500_at   | 838         | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  affy_hg_u133_plus_2 entrezgene\n",
       "1 202763_at           836       \n",
       "2 209310_s_at         837       \n",
       "3 207500_at           838       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "affyids=c(\"202763_at\",\"209310_s_at\",\"207500_at\")\n",
    "getBM(attributes=c('affy_hg_u133_plus_2', 'entrezgene'), \n",
    "      filters = 'affy_hg_u133_plus_2', \n",
    "      values = affyids, \n",
    "      mart = ensembl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
